// @alloralabs/allora-sdk/dist/src/v2/types/emissions/v7/params.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: emissions/v7/params.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "emissions.v7";
function createBaseParams() {
    return {
        version: "",
        maxSerializedMsgLength: "0",
        minTopicWeight: "",
        requiredMinimumStake: "",
        removeStakeDelayWindow: "0",
        minEpochLength: "0",
        betaEntropy: "",
        learningRate: "",
        maxGradientThreshold: "",
        minStakeFraction: "",
        maxUnfulfilledWorkerRequests: "0",
        maxUnfulfilledReputerRequests: "0",
        topicRewardStakeImportance: "",
        topicRewardFeeRevenueImportance: "",
        topicRewardAlpha: "",
        taskRewardAlpha: "",
        validatorsVsAlloraPercentReward: "",
        maxSamplesToScaleScores: "0",
        maxTopInferersToReward: "0",
        maxTopForecastersToReward: "0",
        maxTopReputersToReward: "0",
        createTopicFee: "",
        gradientDescentMaxIters: "0",
        registrationFee: "",
        defaultPageLimit: "0",
        maxPageLimit: "0",
        minEpochLengthRecordLimit: "0",
        blocksPerMonth: "0",
        pRewardInference: "",
        pRewardForecast: "",
        pRewardReputer: "",
        cRewardInference: "",
        cRewardForecast: "",
        cNorm: "",
        epsilonReputer: "",
        halfMaxProcessStakeRemovalsEndBlock: "0",
        epsilonSafeDiv: "",
        dataSendingFee: "",
        maxElementsPerForecast: "0",
        maxActiveTopicsPerBlock: "0",
        maxStringLength: "0",
        initialRegretQuantile: "",
        pNormSafeDiv: "",
        globalWhitelistEnabled: false,
        topicCreatorWhitelistEnabled: false,
        minExperiencedWorkerRegrets: "0",
        inferenceOutlierDetectionThreshold: "",
        inferenceOutlierDetectionAlpha: "",
        lambdaInitialScore: "",
        globalWorkerWhitelistEnabled: false,
        globalReputerWhitelistEnabled: false,
        globalAdminWhitelistAppended: false,
        maxWhitelistInputArrayLength: "0",
    };
}
exports.Params = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.version !== "") {
            writer.uint32(10).string(message.version);
        }
        if (message.maxSerializedMsgLength !== "0") {
            writer.uint32(16).int64(message.maxSerializedMsgLength);
        }
        if (message.minTopicWeight !== "") {
            writer.uint32(26).string(message.minTopicWeight);
        }
        if (message.requiredMinimumStake !== "") {
            writer.uint32(42).string(message.requiredMinimumStake);
        }
        if (message.removeStakeDelayWindow !== "0") {
            writer.uint32(48).int64(message.removeStakeDelayWindow);
        }
        if (message.minEpochLength !== "0") {
            writer.uint32(56).int64(message.minEpochLength);
        }
        if (message.betaEntropy !== "") {
            writer.uint32(66).string(message.betaEntropy);
        }
        if (message.learningRate !== "") {
            writer.uint32(74).string(message.learningRate);
        }
        if (message.maxGradientThreshold !== "") {
            writer.uint32(82).string(message.maxGradientThreshold);
        }
        if (message.minStakeFraction !== "") {
            writer.uint32(90).string(message.minStakeFraction);
        }
        if (message.maxUnfulfilledWorkerRequests !== "0") {
            writer.uint32(104).uint64(message.maxUnfulfilledWorkerRequests);
        }
        if (message.maxUnfulfilledReputerRequests !== "0") {
            writer.uint32(112).uint64(message.maxUnfulfilledReputerRequests);
        }
        if (message.topicRewardStakeImportance !== "") {
            writer.uint32(122).string(message.topicRewardStakeImportance);
        }
        if (message.topicRewardFeeRevenueImportance !== "") {
            writer.uint32(130).string(message.topicRewardFeeRevenueImportance);
        }
        if (message.topicRewardAlpha !== "") {
            writer.uint32(138).string(message.topicRewardAlpha);
        }
        if (message.taskRewardAlpha !== "") {
            writer.uint32(146).string(message.taskRewardAlpha);
        }
        if (message.validatorsVsAlloraPercentReward !== "") {
            writer.uint32(154).string(message.validatorsVsAlloraPercentReward);
        }
        if (message.maxSamplesToScaleScores !== "0") {
            writer.uint32(160).uint64(message.maxSamplesToScaleScores);
        }
        if (message.maxTopInferersToReward !== "0") {
            writer.uint32(168).uint64(message.maxTopInferersToReward);
        }
        if (message.maxTopForecastersToReward !== "0") {
            writer.uint32(176).uint64(message.maxTopForecastersToReward);
        }
        if (message.maxTopReputersToReward !== "0") {
            writer.uint32(184).uint64(message.maxTopReputersToReward);
        }
        if (message.createTopicFee !== "") {
            writer.uint32(194).string(message.createTopicFee);
        }
        if (message.gradientDescentMaxIters !== "0") {
            writer.uint32(200).uint64(message.gradientDescentMaxIters);
        }
        if (message.registrationFee !== "") {
            writer.uint32(226).string(message.registrationFee);
        }
        if (message.defaultPageLimit !== "0") {
            writer.uint32(232).uint64(message.defaultPageLimit);
        }
        if (message.maxPageLimit !== "0") {
            writer.uint32(240).uint64(message.maxPageLimit);
        }
        if (message.minEpochLengthRecordLimit !== "0") {
            writer.uint32(248).int64(message.minEpochLengthRecordLimit);
        }
        if (message.blocksPerMonth !== "0") {
            writer.uint32(256).uint64(message.blocksPerMonth);
        }
        if (message.pRewardInference !== "") {
            writer.uint32(266).string(message.pRewardInference);
        }
        if (message.pRewardForecast !== "") {
            writer.uint32(274).string(message.pRewardForecast);
        }
        if (message.pRewardReputer !== "") {
            writer.uint32(282).string(message.pRewardReputer);
        }
        if (message.cRewardInference !== "") {
            writer.uint32(290).string(message.cRewardInference);
        }
        if (message.cRewardForecast !== "") {
            writer.uint32(298).string(message.cRewardForecast);
        }
        if (message.cNorm !== "") {
            writer.uint32(306).string(message.cNorm);
        }
        if (message.epsilonReputer !== "") {
            writer.uint32(322).string(message.epsilonReputer);
        }
        if (message.halfMaxProcessStakeRemovalsEndBlock !== "0") {
            writer.uint32(336).uint64(message.halfMaxProcessStakeRemovalsEndBlock);
        }
        if (message.epsilonSafeDiv !== "") {
            writer.uint32(346).string(message.epsilonSafeDiv);
        }
        if (message.dataSendingFee !== "") {
            writer.uint32(354).string(message.dataSendingFee);
        }
        if (message.maxElementsPerForecast !== "0") {
            writer.uint32(360).uint64(message.maxElementsPerForecast);
        }
        if (message.maxActiveTopicsPerBlock !== "0") {
            writer.uint32(368).uint64(message.maxActiveTopicsPerBlock);
        }
        if (message.maxStringLength !== "0") {
            writer.uint32(376).uint64(message.maxStringLength);
        }
        if (message.initialRegretQuantile !== "") {
            writer.uint32(386).string(message.initialRegretQuantile);
        }
        if (message.pNormSafeDiv !== "") {
            writer.uint32(394).string(message.pNormSafeDiv);
        }
        if (message.globalWhitelistEnabled !== false) {
            writer.uint32(400).bool(message.globalWhitelistEnabled);
        }
        if (message.topicCreatorWhitelistEnabled !== false) {
            writer.uint32(408).bool(message.topicCreatorWhitelistEnabled);
        }
        if (message.minExperiencedWorkerRegrets !== "0") {
            writer.uint32(416).uint64(message.minExperiencedWorkerRegrets);
        }
        if (message.inferenceOutlierDetectionThreshold !== "") {
            writer.uint32(426).string(message.inferenceOutlierDetectionThreshold);
        }
        if (message.inferenceOutlierDetectionAlpha !== "") {
            writer.uint32(434).string(message.inferenceOutlierDetectionAlpha);
        }
        if (message.lambdaInitialScore !== "") {
            writer.uint32(442).string(message.lambdaInitialScore);
        }
        if (message.globalWorkerWhitelistEnabled !== false) {
            writer.uint32(448).bool(message.globalWorkerWhitelistEnabled);
        }
        if (message.globalReputerWhitelistEnabled !== false) {
            writer.uint32(456).bool(message.globalReputerWhitelistEnabled);
        }
        if (message.globalAdminWhitelistAppended !== false) {
            writer.uint32(464).bool(message.globalAdminWhitelistAppended);
        }
        if (message.maxWhitelistInputArrayLength !== "0") {
            writer.uint32(472).uint64(message.maxWhitelistInputArrayLength);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.version = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.maxSerializedMsgLength = reader.int64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.minTopicWeight = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.requiredMinimumStake = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.removeStakeDelayWindow = reader.int64().toString();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.minEpochLength = reader.int64().toString();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.betaEntropy = reader.string();
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.learningRate = reader.string();
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.maxGradientThreshold = reader.string();
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.minStakeFraction = reader.string();
                    continue;
                }
                case 13: {
                    if (tag !== 104) {
                        break;
                    }
                    message.maxUnfulfilledWorkerRequests = reader.uint64().toString();
                    continue;
                }
                case 14: {
                    if (tag !== 112) {
                        break;
                    }
                    message.maxUnfulfilledReputerRequests = reader.uint64().toString();
                    continue;
                }
                case 15: {
                    if (tag !== 122) {
                        break;
                    }
                    message.topicRewardStakeImportance = reader.string();
                    continue;
                }
                case 16: {
                    if (tag !== 130) {
                        break;
                    }
                    message.topicRewardFeeRevenueImportance = reader.string();
                    continue;
                }
                case 17: {
                    if (tag !== 138) {
                        break;
                    }
                    message.topicRewardAlpha = reader.string();
                    continue;
                }
                case 18: {
                    if (tag !== 146) {
                        break;
                    }
                    message.taskRewardAlpha = reader.string();
                    continue;
                }
                case 19: {
                    if (tag !== 154) {
                        break;
                    }
                    message.validatorsVsAlloraPercentReward = reader.string();
                    continue;
                }
                case 20: {
                    if (tag !== 160) {
                        break;
                    }
                    message.maxSamplesToScaleScores = reader.uint64().toString();
                    continue;
                }
                case 21: {
                    if (tag !== 168) {
                        break;
                    }
                    message.maxTopInferersToReward = reader.uint64().toString();
                    continue;
                }
                case 22: {
                    if (tag !== 176) {
                        break;
                    }
                    message.maxTopForecastersToReward = reader.uint64().toString();
                    continue;
                }
                case 23: {
                    if (tag !== 184) {
                        break;
                    }
                    message.maxTopReputersToReward = reader.uint64().toString();
                    continue;
                }
                case 24: {
                    if (tag !== 194) {
                        break;
                    }
                    message.createTopicFee = reader.string();
                    continue;
                }
                case 25: {
                    if (tag !== 200) {
                        break;
                    }
                    message.gradientDescentMaxIters = reader.uint64().toString();
                    continue;
                }
                case 28: {
                    if (tag !== 226) {
                        break;
                    }
                    message.registrationFee = reader.string();
                    continue;
                }
                case 29: {
                    if (tag !== 232) {
                        break;
                    }
                    message.defaultPageLimit = reader.uint64().toString();
                    continue;
                }
                case 30: {
                    if (tag !== 240) {
                        break;
                    }
                    message.maxPageLimit = reader.uint64().toString();
                    continue;
                }
                case 31: {
                    if (tag !== 248) {
                        break;
                    }
                    message.minEpochLengthRecordLimit = reader.int64().toString();
                    continue;
                }
                case 32: {
                    if (tag !== 256) {
                        break;
                    }
                    message.blocksPerMonth = reader.uint64().toString();
                    continue;
                }
                case 33: {
                    if (tag !== 266) {
                        break;
                    }
                    message.pRewardInference = reader.string();
                    continue;
                }
                case 34: {
                    if (tag !== 274) {
                        break;
                    }
                    message.pRewardForecast = reader.string();
                    continue;
                }
                case 35: {
                    if (tag !== 282) {
                        break;
                    }
                    message.pRewardReputer = reader.string();
                    continue;
                }
                case 36: {
                    if (tag !== 290) {
                        break;
                    }
                    message.cRewardInference = reader.string();
                    continue;
                }
                case 37: {
                    if (tag !== 298) {
                        break;
                    }
                    message.cRewardForecast = reader.string();
                    continue;
                }
                case 38: {
                    if (tag !== 306) {
                        break;
                    }
                    message.cNorm = reader.string();
                    continue;
                }
                case 40: {
                    if (tag !== 322) {
                        break;
                    }
                    message.epsilonReputer = reader.string();
                    continue;
                }
                case 42: {
                    if (tag !== 336) {
                        break;
                    }
                    message.halfMaxProcessStakeRemovalsEndBlock = reader.uint64().toString();
                    continue;
                }
                case 43: {
                    if (tag !== 346) {
                        break;
                    }
                    message.epsilonSafeDiv = reader.string();
                    continue;
                }
                case 44: {
                    if (tag !== 354) {
                        break;
                    }
                    message.dataSendingFee = reader.string();
                    continue;
                }
                case 45: {
                    if (tag !== 360) {
                        break;
                    }
                    message.maxElementsPerForecast = reader.uint64().toString();
                    continue;
                }
                case 46: {
                    if (tag !== 368) {
                        break;
                    }
                    message.maxActiveTopicsPerBlock = reader.uint64().toString();
                    continue;
                }
                case 47: {
                    if (tag !== 376) {
                        break;
                    }
                    message.maxStringLength = reader.uint64().toString();
                    continue;
                }
                case 48: {
                    if (tag !== 386) {
                        break;
                    }
                    message.initialRegretQuantile = reader.string();
                    continue;
                }
                case 49: {
                    if (tag !== 394) {
                        break;
                    }
                    message.pNormSafeDiv = reader.string();
                    continue;
                }
                case 50: {
                    if (tag !== 400) {
                        break;
                    }
                    message.globalWhitelistEnabled = reader.bool();
                    continue;
                }
                case 51: {
                    if (tag !== 408) {
                        break;
                    }
                    message.topicCreatorWhitelistEnabled = reader.bool();
                    continue;
                }
                case 52: {
                    if (tag !== 416) {
                        break;
                    }
                    message.minExperiencedWorkerRegrets = reader.uint64().toString();
                    continue;
                }
                case 53: {
                    if (tag !== 426) {
                        break;
                    }
                    message.inferenceOutlierDetectionThreshold = reader.string();
                    continue;
                }
                case 54: {
                    if (tag !== 434) {
                        break;
                    }
                    message.inferenceOutlierDetectionAlpha = reader.string();
                    continue;
                }
                case 55: {
                    if (tag !== 442) {
                        break;
                    }
                    message.lambdaInitialScore = reader.string();
                    continue;
                }
                case 56: {
                    if (tag !== 448) {
                        break;
                    }
                    message.globalWorkerWhitelistEnabled = reader.bool();
                    continue;
                }
                case 57: {
                    if (tag !== 456) {
                        break;
                    }
                    message.globalReputerWhitelistEnabled = reader.bool();
                    continue;
                }
                case 58: {
                    if (tag !== 464) {
                        break;
                    }
                    message.globalAdminWhitelistAppended = reader.bool();
                    continue;
                }
                case 59: {
                    if (tag !== 472) {
                        break;
                    }
                    message.maxWhitelistInputArrayLength = reader.uint64().toString();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            version: isSet(object.version) ? globalThis.String(object.version) : "",
            maxSerializedMsgLength: isSet(object.maxSerializedMsgLength)
                ? globalThis.String(object.maxSerializedMsgLength)
                : "0",
            minTopicWeight: isSet(object.minTopicWeight) ? globalThis.String(object.minTopicWeight) : "",
            requiredMinimumStake: isSet(object.requiredMinimumStake) ? globalThis.String(object.requiredMinimumStake) : "",
            removeStakeDelayWindow: isSet(object.removeStakeDelayWindow)
                ? globalThis.String(object.removeStakeDelayWindow)
                : "0",
            minEpochLength: isSet(object.minEpochLength) ? globalThis.String(object.minEpochLength) : "0",
            betaEntropy: isSet(object.betaEntropy) ? globalThis.String(object.betaEntropy) : "",
            learningRate: isSet(object.learningRate) ? globalThis.String(object.learningRate) : "",
            maxGradientThreshold: isSet(object.maxGradientThreshold) ? globalThis.String(object.maxGradientThreshold) : "",
            minStakeFraction: isSet(object.minStakeFraction) ? globalThis.String(object.minStakeFraction) : "",
            maxUnfulfilledWorkerRequests: isSet(object.maxUnfulfilledWorkerRequests)
                ? globalThis.String(object.maxUnfulfilledWorkerRequests)
                : "0",
            maxUnfulfilledReputerRequests: isSet(object.maxUnfulfilledReputerRequests)
                ? globalThis.String(object.maxUnfulfilledReputerRequests)
                : "0",
            topicRewardStakeImportance: isSet(object.topicRewardStakeImportance)
                ? globalThis.String(object.topicRewardStakeImportance)
                : "",
            topicRewardFeeRevenueImportance: isSet(object.topicRewardFeeRevenueImportance)
                ? globalThis.String(object.topicRewardFeeRevenueImportance)
                : "",
            topicRewardAlpha: isSet(object.topicRewardAlpha) ? globalThis.String(object.topicRewardAlpha) : "",
            taskRewardAlpha: isSet(object.taskRewardAlpha) ? globalThis.String(object.taskRewardAlpha) : "",
            validatorsVsAlloraPercentReward: isSet(object.validatorsVsAlloraPercentReward)
                ? globalThis.String(object.validatorsVsAlloraPercentReward)
                : "",
            maxSamplesToScaleScores: isSet(object.maxSamplesToScaleScores)
                ? globalThis.String(object.maxSamplesToScaleScores)
                : "0",
            maxTopInferersToReward: isSet(object.maxTopInferersToReward)
                ? globalThis.String(object.maxTopInferersToReward)
                : "0",
            maxTopForecastersToReward: isSet(object.maxTopForecastersToReward)
                ? globalThis.String(object.maxTopForecastersToReward)
                : "0",
            maxTopReputersToReward: isSet(object.maxTopReputersToReward)
                ? globalThis.String(object.maxTopReputersToReward)
                : "0",
            createTopicFee: isSet(object.createTopicFee) ? globalThis.String(object.createTopicFee) : "",
            gradientDescentMaxIters: isSet(object.gradientDescentMaxIters)
                ? globalThis.String(object.gradientDescentMaxIters)
                : "0",
            registrationFee: isSet(object.registrationFee) ? globalThis.String(object.registrationFee) : "",
            defaultPageLimit: isSet(object.defaultPageLimit) ? globalThis.String(object.defaultPageLimit) : "0",
            maxPageLimit: isSet(object.maxPageLimit) ? globalThis.String(object.maxPageLimit) : "0",
            minEpochLengthRecordLimit: isSet(object.minEpochLengthRecordLimit)
                ? globalThis.String(object.minEpochLengthRecordLimit)
                : "0",
            blocksPerMonth: isSet(object.blocksPerMonth) ? globalThis.String(object.blocksPerMonth) : "0",
            pRewardInference: isSet(object.pRewardInference) ? globalThis.String(object.pRewardInference) : "",
            pRewardForecast: isSet(object.pRewardForecast) ? globalThis.String(object.pRewardForecast) : "",
            pRewardReputer: isSet(object.pRewardReputer) ? globalThis.String(object.pRewardReputer) : "",
            cRewardInference: isSet(object.cRewardInference) ? globalThis.String(object.cRewardInference) : "",
            cRewardForecast: isSet(object.cRewardForecast) ? globalThis.String(object.cRewardForecast) : "",
            cNorm: isSet(object.cNorm) ? globalThis.String(object.cNorm) : "",
            epsilonReputer: isSet(object.epsilonReputer) ? globalThis.String(object.epsilonReputer) : "",
            halfMaxProcessStakeRemovalsEndBlock: isSet(object.halfMaxProcessStakeRemovalsEndBlock)
                ? globalThis.String(object.halfMaxProcessStakeRemovalsEndBlock)
                : "0",
            epsilonSafeDiv: isSet(object.epsilonSafeDiv) ? globalThis.String(object.epsilonSafeDiv) : "",
            dataSendingFee: isSet(object.dataSendingFee) ? globalThis.String(object.dataSendingFee) : "",
            maxElementsPerForecast: isSet(object.maxElementsPerForecast)
                ? globalThis.String(object.maxElementsPerForecast)
                : "0",
            maxActiveTopicsPerBlock: isSet(object.maxActiveTopicsPerBlock)
                ? globalThis.String(object.maxActiveTopicsPerBlock)
                : "0",
            maxStringLength: isSet(object.maxStringLength) ? globalThis.String(object.maxStringLength) : "0",
            initialRegretQuantile: isSet(object.initialRegretQuantile) ? globalThis.String(object.initialRegretQuantile) : "",
            pNormSafeDiv: isSet(object.pNormSafeDiv) ? globalThis.String(object.pNormSafeDiv) : "",
            globalWhitelistEnabled: isSet(object.globalWhitelistEnabled)
                ? globalThis.Boolean(object.globalWhitelistEnabled)
                : false,
            topicCreatorWhitelistEnabled: isSet(object.topicCreatorWhitelistEnabled)
                ? globalThis.Boolean(object.topicCreatorWhitelistEnabled)
                : false,
            minExperiencedWorkerRegrets: isSet(object.minExperiencedWorkerRegrets)
                ? globalThis.String(object.minExperiencedWorkerRegrets)
                : "0",
            inferenceOutlierDetectionThreshold: isSet(object.inferenceOutlierDetectionThreshold)
                ? globalThis.String(object.inferenceOutlierDetectionThreshold)
                : "",
            inferenceOutlierDetectionAlpha: isSet(object.inferenceOutlierDetectionAlpha)
                ? globalThis.String(object.inferenceOutlierDetectionAlpha)
                : "",
            lambdaInitialScore: isSet(object.lambdaInitialScore) ? globalThis.String(object.lambdaInitialScore) : "",
            globalWorkerWhitelistEnabled: isSet(object.globalWorkerWhitelistEnabled)
                ? globalThis.Boolean(object.globalWorkerWhitelistEnabled)
                : false,
            globalReputerWhitelistEnabled: isSet(object.globalReputerWhitelistEnabled)
                ? globalThis.Boolean(object.globalReputerWhitelistEnabled)
                : false,
            globalAdminWhitelistAppended: isSet(object.globalAdminWhitelistAppended)
                ? globalThis.Boolean(object.globalAdminWhitelistAppended)
                : false,
            maxWhitelistInputArrayLength: isSet(object.maxWhitelistInputArrayLength)
                ? globalThis.String(object.maxWhitelistInputArrayLength)
                : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.version !== "") {
            obj.version = message.version;
        }
        if (message.maxSerializedMsgLength !== "0") {
            obj.maxSerializedMsgLength = message.maxSerializedMsgLength;
        }
        if (message.minTopicWeight !== "") {
            obj.minTopicWeight = message.minTopicWeight;
        }
        if (message.requiredMinimumStake !== "") {
            obj.requiredMinimumStake = message.requiredMinimumStake;
        }
        if (message.removeStakeDelayWindow !== "0") {
            obj.removeStakeDelayWindow = message.removeStakeDelayWindow;
        }
        if (message.minEpochLength !== "0") {
            obj.minEpochLength = message.minEpochLength;
        }
        if (message.betaEntropy !== "") {
            obj.betaEntropy = message.betaEntropy;
        }
        if (message.learningRate !== "") {
            obj.learningRate = message.learningRate;
        }
        if (message.maxGradientThreshold !== "") {
            obj.maxGradientThreshold = message.maxGradientThreshold;
        }
        if (message.minStakeFraction !== "") {
            obj.minStakeFraction = message.minStakeFraction;
        }
        if (message.maxUnfulfilledWorkerRequests !== "0") {
            obj.maxUnfulfilledWorkerRequests = message.maxUnfulfilledWorkerRequests;
        }
        if (message.maxUnfulfilledReputerRequests !== "0") {
            obj.maxUnfulfilledReputerRequests = message.maxUnfulfilledReputerRequests;
        }
        if (message.topicRewardStakeImportance !== "") {
            obj.topicRewardStakeImportance = message.topicRewardStakeImportance;
        }
        if (message.topicRewardFeeRevenueImportance !== "") {
            obj.topicRewardFeeRevenueImportance = message.topicRewardFeeRevenueImportance;
        }
        if (message.topicRewardAlpha !== "") {
            obj.topicRewardAlpha = message.topicRewardAlpha;
        }
        if (message.taskRewardAlpha !== "") {
            obj.taskRewardAlpha = message.taskRewardAlpha;
        }
        if (message.validatorsVsAlloraPercentReward !== "") {
            obj.validatorsVsAlloraPercentReward = message.validatorsVsAlloraPercentReward;
        }
        if (message.maxSamplesToScaleScores !== "0") {
            obj.maxSamplesToScaleScores = message.maxSamplesToScaleScores;
        }
        if (message.maxTopInferersToReward !== "0") {
            obj.maxTopInferersToReward = message.maxTopInferersToReward;
        }
        if (message.maxTopForecastersToReward !== "0") {
            obj.maxTopForecastersToReward = message.maxTopForecastersToReward;
        }
        if (message.maxTopReputersToReward !== "0") {
            obj.maxTopReputersToReward = message.maxTopReputersToReward;
        }
        if (message.createTopicFee !== "") {
            obj.createTopicFee = message.createTopicFee;
        }
        if (message.gradientDescentMaxIters !== "0") {
            obj.gradientDescentMaxIters = message.gradientDescentMaxIters;
        }
        if (message.registrationFee !== "") {
            obj.registrationFee = message.registrationFee;
        }
        if (message.defaultPageLimit !== "0") {
            obj.defaultPageLimit = message.defaultPageLimit;
        }
        if (message.maxPageLimit !== "0") {
            obj.maxPageLimit = message.maxPageLimit;
        }
        if (message.minEpochLengthRecordLimit !== "0") {
            obj.minEpochLengthRecordLimit = message.minEpochLengthRecordLimit;
        }
        if (message.blocksPerMonth !== "0") {
            obj.blocksPerMonth = message.blocksPerMonth;
        }
        if (message.pRewardInference !== "") {
            obj.pRewardInference = message.pRewardInference;
        }
        if (message.pRewardForecast !== "") {
            obj.pRewardForecast = message.pRewardForecast;
        }
        if (message.pRewardReputer !== "") {
            obj.pRewardReputer = message.pRewardReputer;
        }
        if (message.cRewardInference !== "") {
            obj.cRewardInference = message.cRewardInference;
        }
        if (message.cRewardForecast !== "") {
            obj.cRewardForecast = message.cRewardForecast;
        }
        if (message.cNorm !== "") {
            obj.cNorm = message.cNorm;
        }
        if (message.epsilonReputer !== "") {
            obj.epsilonReputer = message.epsilonReputer;
        }
        if (message.halfMaxProcessStakeRemovalsEndBlock !== "0") {
            obj.halfMaxProcessStakeRemovalsEndBlock = message.halfMaxProcessStakeRemovalsEndBlock;
        }
        if (message.epsilonSafeDiv !== "") {
            obj.epsilonSafeDiv = message.epsilonSafeDiv;
        }
        if (message.dataSendingFee !== "") {
            obj.dataSendingFee = message.dataSendingFee;
        }
        if (message.maxElementsPerForecast !== "0") {
            obj.maxElementsPerForecast = message.maxElementsPerForecast;
        }
        if (message.maxActiveTopicsPerBlock !== "0") {
            obj.maxActiveTopicsPerBlock = message.maxActiveTopicsPerBlock;
        }
        if (message.maxStringLength !== "0") {
            obj.maxStringLength = message.maxStringLength;
        }
        if (message.initialRegretQuantile !== "") {
            obj.initialRegretQuantile = message.initialRegretQuantile;
        }
        if (message.pNormSafeDiv !== "") {
            obj.pNormSafeDiv = message.pNormSafeDiv;
        }
        if (message.globalWhitelistEnabled !== false) {
            obj.globalWhitelistEnabled = message.globalWhitelistEnabled;
        }
        if (message.topicCreatorWhitelistEnabled !== false) {
            obj.topicCreatorWhitelistEnabled = message.topicCreatorWhitelistEnabled;
        }
        if (message.minExperiencedWorkerRegrets !== "0") {
            obj.minExperiencedWorkerRegrets = message.minExperiencedWorkerRegrets;
        }
        if (message.inferenceOutlierDetectionThreshold !== "") {
            obj.inferenceOutlierDetectionThreshold = message.inferenceOutlierDetectionThreshold;
        }
        if (message.inferenceOutlierDetectionAlpha !== "") {
            obj.inferenceOutlierDetectionAlpha = message.inferenceOutlierDetectionAlpha;
        }
        if (message.lambdaInitialScore !== "") {
            obj.lambdaInitialScore = message.lambdaInitialScore;
        }
        if (message.globalWorkerWhitelistEnabled !== false) {
            obj.globalWorkerWhitelistEnabled = message.globalWorkerWhitelistEnabled;
        }
        if (message.globalReputerWhitelistEnabled !== false) {
            obj.globalReputerWhitelistEnabled = message.globalReputerWhitelistEnabled;
        }
        if (message.globalAdminWhitelistAppended !== false) {
            obj.globalAdminWhitelistAppended = message.globalAdminWhitelistAppended;
        }
        if (message.maxWhitelistInputArrayLength !== "0") {
            obj.maxWhitelistInputArrayLength = message.maxWhitelistInputArrayLength;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28;
        const message = createBaseParams();
        message.version = (_a = object.version) !== null && _a !== void 0 ? _a : "";
        message.maxSerializedMsgLength = (_b = object.maxSerializedMsgLength) !== null && _b !== void 0 ? _b : "0";
        message.minTopicWeight = (_c = object.minTopicWeight) !== null && _c !== void 0 ? _c : "";
        message.requiredMinimumStake = (_d = object.requiredMinimumStake) !== null && _d !== void 0 ? _d : "";
        message.removeStakeDelayWindow = (_e = object.removeStakeDelayWindow) !== null && _e !== void 0 ? _e : "0";
        message.minEpochLength = (_f = object.minEpochLength) !== null && _f !== void 0 ? _f : "0";
        message.betaEntropy = (_g = object.betaEntropy) !== null && _g !== void 0 ? _g : "";
        message.learningRate = (_h = object.learningRate) !== null && _h !== void 0 ? _h : "";
        message.maxGradientThreshold = (_j = object.maxGradientThreshold) !== null && _j !== void 0 ? _j : "";
        message.minStakeFraction = (_k = object.minStakeFraction) !== null && _k !== void 0 ? _k : "";
        message.maxUnfulfilledWorkerRequests = (_l = object.maxUnfulfilledWorkerRequests) !== null && _l !== void 0 ? _l : "0";
        message.maxUnfulfilledReputerRequests = (_m = object.maxUnfulfilledReputerRequests) !== null && _m !== void 0 ? _m : "0";
        message.topicRewardStakeImportance = (_o = object.topicRewardStakeImportance) !== null && _o !== void 0 ? _o : "";
        message.topicRewardFeeRevenueImportance = (_p = object.topicRewardFeeRevenueImportance) !== null && _p !== void 0 ? _p : "";
        message.topicRewardAlpha = (_q = object.topicRewardAlpha) !== null && _q !== void 0 ? _q : "";
        message.taskRewardAlpha = (_r = object.taskRewardAlpha) !== null && _r !== void 0 ? _r : "";
        message.validatorsVsAlloraPercentReward = (_s = object.validatorsVsAlloraPercentReward) !== null && _s !== void 0 ? _s : "";
        message.maxSamplesToScaleScores = (_t = object.maxSamplesToScaleScores) !== null && _t !== void 0 ? _t : "0";
        message.maxTopInferersToReward = (_u = object.maxTopInferersToReward) !== null && _u !== void 0 ? _u : "0";
        message.maxTopForecastersToReward = (_v = object.maxTopForecastersToReward) !== null && _v !== void 0 ? _v : "0";
        message.maxTopReputersToReward = (_w = object.maxTopReputersToReward) !== null && _w !== void 0 ? _w : "0";
        message.createTopicFee = (_x = object.createTopicFee) !== null && _x !== void 0 ? _x : "";
        message.gradientDescentMaxIters = (_y = object.gradientDescentMaxIters) !== null && _y !== void 0 ? _y : "0";
        message.registrationFee = (_z = object.registrationFee) !== null && _z !== void 0 ? _z : "";
        message.defaultPageLimit = (_0 = object.defaultPageLimit) !== null && _0 !== void 0 ? _0 : "0";
        message.maxPageLimit = (_1 = object.maxPageLimit) !== null && _1 !== void 0 ? _1 : "0";
        message.minEpochLengthRecordLimit = (_2 = object.minEpochLengthRecordLimit) !== null && _2 !== void 0 ? _2 : "0";
        message.blocksPerMonth = (_3 = object.blocksPerMonth) !== null && _3 !== void 0 ? _3 : "0";
        message.pRewardInference = (_4 = object.pRewardInference) !== null && _4 !== void 0 ? _4 : "";
        message.pRewardForecast = (_5 = object.pRewardForecast) !== null && _5 !== void 0 ? _5 : "";
        message.pRewardReputer = (_6 = object.pRewardReputer) !== null && _6 !== void 0 ? _6 : "";
        message.cRewardInference = (_7 = object.cRewardInference) !== null && _7 !== void 0 ? _7 : "";
        message.cRewardForecast = (_8 = object.cRewardForecast) !== null && _8 !== void 0 ? _8 : "";
        message.cNorm = (_9 = object.cNorm) !== null && _9 !== void 0 ? _9 : "";
        message.epsilonReputer = (_10 = object.epsilonReputer) !== null && _10 !== void 0 ? _10 : "";
        message.halfMaxProcessStakeRemovalsEndBlock = (_11 = object.halfMaxProcessStakeRemovalsEndBlock) !== null && _11 !== void 0 ? _11 : "0";
        message.epsilonSafeDiv = (_12 = object.epsilonSafeDiv) !== null && _12 !== void 0 ? _12 : "";
        message.dataSendingFee = (_13 = object.dataSendingFee) !== null && _13 !== void 0 ? _13 : "";
        message.maxElementsPerForecast = (_14 = object.maxElementsPerForecast) !== null && _14 !== void 0 ? _14 : "0";
        message.maxActiveTopicsPerBlock = (_15 = object.maxActiveTopicsPerBlock) !== null && _15 !== void 0 ? _15 : "0";
        message.maxStringLength = (_16 = object.maxStringLength) !== null && _16 !== void 0 ? _16 : "0";
        message.initialRegretQuantile = (_17 = object.initialRegretQuantile) !== null && _17 !== void 0 ? _17 : "";
        message.pNormSafeDiv = (_18 = object.pNormSafeDiv) !== null && _18 !== void 0 ? _18 : "";
        message.globalWhitelistEnabled = (_19 = object.globalWhitelistEnabled) !== null && _19 !== void 0 ? _19 : false;
        message.topicCreatorWhitelistEnabled = (_20 = object.topicCreatorWhitelistEnabled) !== null && _20 !== void 0 ? _20 : false;
        message.minExperiencedWorkerRegrets = (_21 = object.minExperiencedWorkerRegrets) !== null && _21 !== void 0 ? _21 : "0";
        message.inferenceOutlierDetectionThreshold = (_22 = object.inferenceOutlierDetectionThreshold) !== null && _22 !== void 0 ? _22 : "";
        message.inferenceOutlierDetectionAlpha = (_23 = object.inferenceOutlierDetectionAlpha) !== null && _23 !== void 0 ? _23 : "";
        message.lambdaInitialScore = (_24 = object.lambdaInitialScore) !== null && _24 !== void 0 ? _24 : "";
        message.globalWorkerWhitelistEnabled = (_25 = object.globalWorkerWhitelistEnabled) !== null && _25 !== void 0 ? _25 : false;
        message.globalReputerWhitelistEnabled = (_26 = object.globalReputerWhitelistEnabled) !== null && _26 !== void 0 ? _26 : false;
        message.globalAdminWhitelistAppended = (_27 = object.globalAdminWhitelistAppended) !== null && _27 !== void 0 ? _27 : false;
        message.maxWhitelistInputArrayLength = (_28 = object.maxWhitelistInputArrayLength) !== null && _28 !== void 0 ? _28 : "0";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
