// @zerodev/intent/_cjs/actions/prepareUserIntent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareUserIntent = prepareUserIntent;
const sdk_1 = require("@zerodev/sdk");
const viem_1 = require("viem");
const utils_1 = require("viem/utils");
const getIntent_js_1 = require("./getIntent.js");
async function prepareUserIntent(client, parameters, version) {
    const { account: account_ = client.account } = parameters;
    if (!account_)
        throw new sdk_1.AccountNotFoundError();
    const account = (0, utils_1.parseAccount)(account_);
    const { inputTokens, outputTokens, chainId, gasToken } = parameters;
    const callData = await (async () => {
        if (parameters.calls)
            return account.encodeCalls(parameters.calls.map((call_) => {
                const call = call_;
                if ("abi" in call)
                    return {
                        data: (0, viem_1.encodeFunctionData)(call),
                        to: call.to,
                        value: call.value,
                    };
                return call;
            }));
        return parameters.callData ?? "0x";
    })();
    const factoryAddress = account.factoryAddress;
    const factoryData = await account.generateInitCode();
    const initData = (0, viem_1.concatHex)([factoryAddress, factoryData]);
    return (0, getIntent_js_1.getIntent)(client, {
        recipient: account.address,
        callData,
        inputTokens: inputTokens ?? [],
        outputTokens: outputTokens ?? [],
        gasToken,
        chainId,
        initData,
    }, version);
}
