// @zerodev/intent/_cjs/utils/installIntentExecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIntentExecutorPluginData = exports.getInstallIntentExecutorCall = exports.installIntentExecutor = void 0;
const sdk_1 = require("@zerodev/sdk");
const viem_1 = require("viem");
const constants_js_1 = require("../config/constants.js");
const installIntentExecutor = (version) => (0, viem_1.encodeFunctionData)({
    abi: sdk_1.KernelV3_1AccountAbi,
    functionName: "installModule",
    args: [
        BigInt(2),
        constants_js_1.IntentVersionToAddressesMap[version].intentExecutorAddress,
        (0, viem_1.concatHex)([
            viem_1.zeroAddress,
            (0, viem_1.encodeAbiParameters)((0, viem_1.parseAbiParameters)(["bytes", "bytes"]), [
                "0x",
                "0x",
            ]),
        ]),
    ],
});
exports.installIntentExecutor = installIntentExecutor;
const getInstallIntentExecutorCall = ({ accountAddress, version, }) => {
    return {
        to: accountAddress,
        data: (0, exports.installIntentExecutor)(version),
        value: 0n,
    };
};
exports.getInstallIntentExecutorCall = getInstallIntentExecutorCall;
const getIntentExecutorPluginData = (version) => {
    return {
        type: 2,
        address: constants_js_1.IntentVersionToAddressesMap[version].intentExecutorAddress,
        data: (0, viem_1.concatHex)([
            viem_1.zeroAddress,
            (0, viem_1.encodeAbiParameters)((0, viem_1.parseAbiParameters)(["bytes", "bytes"]), ["0x", "0x"]),
        ]),
    };
};
exports.getIntentExecutorPluginData = getIntentExecutorPluginData;
