// @alloralabs/allora-sdk/dist/src/v2/types/cosmos/bank/module/v1/module.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: cosmos/bank/module/v1/module.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "cosmos.bank.module.v1";
function createBaseModule() {
    return { blockedModuleAccountsOverride: [], authority: "" };
}
exports.Module = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.blockedModuleAccountsOverride) {
            writer.uint32(10).string(v);
        }
        if (message.authority !== "") {
            writer.uint32(18).string(message.authority);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.blockedModuleAccountsOverride.push(reader.string());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.authority = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            blockedModuleAccountsOverride: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.blockedModuleAccountsOverride)
                ? object.blockedModuleAccountsOverride.map((e) => globalThis.String(e))
                : [],
            authority: isSet(object.authority) ? globalThis.String(object.authority) : "",
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.blockedModuleAccountsOverride) === null || _a === void 0 ? void 0 : _a.length) {
            obj.blockedModuleAccountsOverride = message.blockedModuleAccountsOverride;
        }
        if (message.authority !== "") {
            obj.authority = message.authority;
        }
        return obj;
    },
    create(base) {
        return exports.Module.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseModule();
        message.blockedModuleAccountsOverride = ((_a = object.blockedModuleAccountsOverride) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.authority = (_b = object.authority) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
