// @alloralabs/allora-sdk/dist/src/v2/types/cosmos/tx/config/v1/config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: cosmos/tx/config/v1/config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "cosmos.tx.config.v1";
function createBaseConfig() {
    return { skipAnteHandler: false, skipPostHandler: false };
}
exports.Config = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.skipAnteHandler !== false) {
            writer.uint32(8).bool(message.skipAnteHandler);
        }
        if (message.skipPostHandler !== false) {
            writer.uint32(16).bool(message.skipPostHandler);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.skipAnteHandler = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.skipPostHandler = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            skipAnteHandler: isSet(object.skipAnteHandler) ? globalThis.Boolean(object.skipAnteHandler) : false,
            skipPostHandler: isSet(object.skipPostHandler) ? globalThis.Boolean(object.skipPostHandler) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.skipAnteHandler !== false) {
            obj.skipAnteHandler = message.skipAnteHandler;
        }
        if (message.skipPostHandler !== false) {
            obj.skipPostHandler = message.skipPostHandler;
        }
        return obj;
    },
    create(base) {
        return exports.Config.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseConfig();
        message.skipAnteHandler = (_a = object.skipAnteHandler) !== null && _a !== void 0 ? _a : false;
        message.skipPostHandler = (_b = object.skipPostHandler) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
