// @opensea/seaport-js/lib/typechain-types/factories/seaport-core/src/lib/ZoneInteraction__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneInteraction__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "orderHash",
                type: "bytes32",
            },
        ],
        name: "InvalidContractOrder",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "orderHash",
                type: "bytes32",
            },
        ],
        name: "InvalidRestrictedOrder",
        type: "error",
    },
];
const _bytecode = "0x6080604052348015600e575f80fd5b50603e80601a5f395ff3fe60806040525f80fdfea264697066735822122030b78722acfb68476127f7f246052e6735a8966a534cb4457bd18c966e8dd62964736f6c63430008180033";
const isSuperArgs = (xs) => xs.length > 1;
class ZoneInteraction__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.ZoneInteraction__factory = ZoneInteraction__factory;
ZoneInteraction__factory.bytecode = _bytecode;
ZoneInteraction__factory.abi = _abi;
