// @opensea/seaport-js/src/typechain-types/factories/seaport-types/src/interfaces/ReentrancyErrors__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  ReentrancyErrors,
  ReentrancyErrorsInterface,
} from "../../../../seaport-types/src/interfaces/ReentrancyErrors";

const _abi = [
  {
    inputs: [],
    name: "NoReentrantCalls",
    type: "error",
  },
  {
    inputs: [],
    name: "TStoreAlreadyActivated",
    type: "error",
  },
  {
    inputs: [],
    name: "TStoreNotSupported",
    type: "error",
  },
  {
    inputs: [],
    name: "TloadTestContractDeploymentFailed",
    type: "error",
  },
] as const;

export class ReentrancyErrors__factory {
  static readonly abi = _abi;
  static createInterface(): ReentrancyErrorsInterface {
    return new Interface(_abi) as ReentrancyErrorsInterface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): ReentrancyErrors {
    return new Contract(address, _abi, runner) as unknown as ReentrancyErrors;
  }
}
