// @coinbase/agentkit/dist/action-providers/allora/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPriceInferenceSchema = exports.GetInferenceByTopicIdSchema = exports.GetAllTopicsSchema = void 0;
const zod_1 = require("zod");
/**
 * Input schema for getting all topics from Allora Network
 */
exports.GetAllTopicsSchema = zod_1.z
    .object({})
    .strip()
    .describe("Instructions for getting all topics from Allora Network");
/**
 * Input schema for getting inference data by topic ID from Allora Network
 */
exports.GetInferenceByTopicIdSchema = zod_1.z
    .object({
    topicId: zod_1.z.number().describe("The ID of the topic to get inference data for"),
})
    .strip()
    .describe("Instructions for getting inference data from Allora Network by topic ID");
/**
 * Input schema for getting price inference for a token/timeframe pair
 */
exports.GetPriceInferenceSchema = zod_1.z
    .object({
    asset: zod_1.z.string().describe("The token to get price inference for (e.g., 'BTC', 'ETH')"),
    timeframe: zod_1.z.string().describe("The timeframe for the prediction (e.g., '8h', '24h')"),
})
    .strip()
    .describe("Instructions for getting price inference for a token/timeframe pair");
