// @zerodev/intent/_cjs/test/tokenRoutes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAINNET_TOKEN_ROUTES = exports.TESTNET_TOKEN_ROUTES = exports.MAINNET_TOKEN_ROUTES_SAME_CHAIN = exports.BSC_TOKEN_ROUTES = exports.BASE_TOKEN_ROUTES = exports.ARB_TOKEN_ROUTES = exports.OP_TOKEN_ROUTES = exports.ETH_TOKEN_ROUTES = exports.BASE_SEPOLIA_TOKEN_ROUTES = exports.SEPOLIA_TOKEN_ROUTES = void 0;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
exports.SEPOLIA_TOKEN_ROUTES = {
    chain: chains_1.sepolia,
    routes: [
        {
            name: "sepolia eth <-> baseSepolia eth",
            inputTokens: [
                {
                    chainId: chains_1.sepolia.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.baseSepolia.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.005", 18),
                },
            ],
        },
        {
            name: "sepolia usdc <-> baseSepolia usdc",
            inputTokens: [
                {
                    chainId: chains_1.sepolia.id,
                    address: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.baseSepolia.id,
                    address: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
                    amount: (0, viem_1.parseUnits)("0.01", 6),
                },
            ],
        },
    ],
};
exports.BASE_SEPOLIA_TOKEN_ROUTES = {
    chain: chains_1.baseSepolia,
    routes: [
        {
            name: "baseSepolia eth <-> sepolia eth",
            inputTokens: [
                {
                    chainId: chains_1.baseSepolia.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.sepolia.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.005", 18),
                },
            ],
        },
        {
            name: "baseSepolia usdc <-> sepolia usdc",
            inputTokens: [
                {
                    chainId: chains_1.baseSepolia.id,
                    address: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.sepolia.id,
                    address: "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238",
                    amount: (0, viem_1.parseUnits)("0.01", 6),
                },
            ],
        },
    ],
};
exports.ETH_TOKEN_ROUTES = {
    chain: chains_1.mainnet,
    routes: [
        {
            name: "mainnet eth <-> op eth",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "mainnet eth <-> arb eth",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "mainnet eth <-> base eth",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "mainnet usdc <-> op usdc",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x4200000000000000000000000000000000000006",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "mainnet usdc <-> arb usdc",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "mainnet usdc <-> base usdc",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "mainnet usdc <-> bsc usdc",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                    amount: (0, viem_1.parseUnits)("1.5", 18),
                },
            ],
        },
        {
            name: "mainnet usdt <-> op usdt",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "mainnet usdt <-> arb usdt",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "mainnet usdt <-> bsc usdt",
            inputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                    amount: (0, viem_1.parseUnits)("1.5", 18),
                },
            ],
        },
    ],
};
exports.OP_TOKEN_ROUTES = {
    chain: chains_1.optimism,
    routes: [
        {
            name: "op eth <-> mainnet eth",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "op eth <-> arb eth",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "op eth <-> base eth",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "op usdc <-> mainnet usdc",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                    amount: (0, viem_1.parseUnits)("500", 6),
                },
            ],
        },
        {
            name: "op usdc <-> arb usdc",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "op usdc <-> base usdc",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                    amount: (0, viem_1.parseUnits)("0.6", 6),
                },
            ],
        },
        {
            name: "op usdt <-> bsc usdc",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                    amount: (0, viem_1.parseUnits)("0.6", 18),
                },
            ],
        },
        {
            name: "op usdt <-> mainnet usdt",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                    amount: (0, viem_1.parseUnits)("500", 6),
                },
            ],
        },
        {
            name: "op usdt <-> arb usdt",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "op usdt <-> bsc usdt",
            inputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                    amount: (0, viem_1.parseUnits)("0.6", 18),
                },
            ],
        },
    ],
};
exports.ARB_TOKEN_ROUTES = {
    chain: chains_1.arbitrum,
    routes: [
        {
            name: "arb eth <-> mainnet eth",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "arb eth <-> op eth",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "arb eth <-> base eth",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "arb usdc <-> mainnet usdc",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                    amount: (0, viem_1.parseUnits)("200", 6),
                },
            ],
        },
        {
            name: "arb usdc <-> op usdc",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
                    amount: (0, viem_1.parseUnits)("0.6", 6),
                },
            ],
        },
        {
            name: "arb usdc <-> base usdc",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                    amount: (0, viem_1.parseUnits)("0.6", 6),
                },
            ],
        },
        {
            name: "arb usdc <-> bsc usdc",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                    amount: (0, viem_1.parseUnits)("0.6", 18),
                },
            ],
        },
        {
            name: "arb usdt <-> mainnet usdt",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                    amount: (0, viem_1.parseUnits)("200", 6),
                },
            ],
        },
        {
            name: "arb usdt <-> op usdt",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                    amount: (0, viem_1.parseUnits)("0.6", 6),
                },
            ],
        },
        {
            name: "arb usdt <-> bsc usdt",
            inputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                    amount: (0, viem_1.parseUnits)("0.6", 18),
                },
            ],
        },
    ],
};
exports.BASE_TOKEN_ROUTES = {
    chain: chains_1.base,
    routes: [
        {
            name: "base eth <-> mainnet eth",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.1", 18),
                },
            ],
        },
        {
            name: "base eth <-> op eth",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "base eth <-> arb eth",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: viem_1.zeroAddress,
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: viem_1.zeroAddress,
                    amount: (0, viem_1.parseUnits)("0.002", 18),
                },
            ],
        },
        {
            name: "base usdc <-> mainnet usdc",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                    amount: (0, viem_1.parseUnits)("200", 6),
                },
            ],
        },
        {
            name: "base usdc <-> op usdc",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "base usdc <-> arb usdc",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "base usdc <-> bsc usdc",
            inputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                    amount: (0, viem_1.parseUnits)("1.5", 18),
                },
            ],
        },
    ],
};
exports.BSC_TOKEN_ROUTES = {
    chain: chains_1.bsc,
    routes: [
        {
            name: "bsc usdc <-> mainnet usdc",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                    amount: (0, viem_1.parseUnits)("200", 6),
                },
            ],
        },
        {
            name: "bsc usdc <-> op usdc",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "bsc usdc <-> arb usdc",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "bsc usdc <-> base usdc",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.base.id,
                    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "bsc usdt <-> mainnet usdt",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.mainnet.id,
                    address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
                    amount: (0, viem_1.parseUnits)("200", 6),
                },
            ],
        },
        {
            name: "bsc usdt <-> op usdt",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.optimism.id,
                    address: "0x94b008aa00579c1307b0ef2c499ad98a8ce58e58",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
        {
            name: "bsc usdt <-> arb usdt",
            inputTokens: [
                {
                    chainId: chains_1.bsc.id,
                    address: "0x55d398326f99059ff775485246999027b3197955",
                },
            ],
            outputTokens: [
                {
                    chainId: chains_1.arbitrum.id,
                    address: "0xfd086bc7cd5c481dcc9c85ebe478a1c0b69fcbb9",
                    amount: (0, viem_1.parseUnits)("1.5", 6),
                },
            ],
        },
    ],
};
exports.MAINNET_TOKEN_ROUTES_SAME_CHAIN = [
    {
        chain: chains_1.mainnet,
        routes: [
            {
                name: "mainnet usdc <-> mainnet usdc",
                inputTokens: [
                    {
                        chainId: chains_1.mainnet.id,
                        address: "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
                    },
                ],
                outputTokens: [],
            },
        ],
    },
    {
        chain: chains_1.optimism,
        routes: [
            {
                name: "optimism usdc <-> optimism usdc",
                inputTokens: [
                    {
                        chainId: chains_1.optimism.id,
                        address: "0x0b2c639c533813f4aa9d7837caf62653d097ff85",
                    },
                ],
                outputTokens: [],
            },
        ],
    },
    {
        chain: chains_1.arbitrum,
        routes: [
            {
                name: "arbitrum usdc <-> arbitrum usdc",
                inputTokens: [
                    {
                        chainId: chains_1.arbitrum.id,
                        address: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
                    },
                ],
                outputTokens: [],
            },
        ],
    },
    {
        chain: chains_1.base,
        routes: [
            {
                name: "base usdc <-> base usdc",
                inputTokens: [
                    {
                        chainId: chains_1.base.id,
                        address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
                    },
                ],
                outputTokens: [],
            },
        ],
    },
    {
        chain: chains_1.bsc,
        routes: [
            {
                name: "bsc usdc <-> bsc usdc",
                inputTokens: [
                    {
                        chainId: chains_1.bsc.id,
                        address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
                    },
                ],
                outputTokens: [],
            },
        ],
    },
];
exports.TESTNET_TOKEN_ROUTES = [
    exports.SEPOLIA_TOKEN_ROUTES,
    exports.BASE_SEPOLIA_TOKEN_ROUTES,
];
exports.MAINNET_TOKEN_ROUTES = [
    exports.ETH_TOKEN_ROUTES,
    exports.OP_TOKEN_ROUTES,
    exports.ARB_TOKEN_ROUTES,
    exports.BASE_TOKEN_ROUTES,
    exports.BSC_TOKEN_ROUTES,
];
