// @across-protocol/app-sdk/dist/utils/typedData.js
import { addressToBytes32 } from "./hex.js";
export function getUpdateDepositTypedData({ signerAddress, originChainId, depositId, updatedMessage, updatedOutputAmount, updatedRecipient, }) {
    return {
        account: signerAddress,
        domain: {
            name: "ACROSS-V2",
            version: "1.0.0",
            chainId: Number(originChainId),
        },
        types: {
            UpdateDepositDetails: [
                { name: "depositId", type: "uint256" },
                { name: "originChainId", type: "uint256" },
                { name: "updatedOutputAmount", type: "uint256" },
                { name: "updatedRecipient", type: "address" },
                { name: "updatedMessage", type: "bytes" },
            ],
        },
        primaryType: "UpdateDepositDetails",
        message: {
            depositId: BigInt(depositId),
            originChainId: BigInt(originChainId),
            updatedOutputAmount,
            updatedRecipient,
            updatedMessage,
        },
    };
}
export function getUpdateDepositTypedDataV3_5({ signerAddress, originChainId, depositId, updatedMessage, updatedOutputAmount, updatedRecipient, }) {
    return {
        account: signerAddress,
        domain: {
            name: "ACROSS-V2",
            version: "1.0.0",
            chainId: Number(originChainId),
        },
        types: {
            UpdateDepositDetails: [
                { name: "depositId", type: "uint256" },
                { name: "originChainId", type: "uint256" },
                { name: "updatedOutputAmount", type: "uint256" },
                { name: "updatedRecipient", type: "bytes32" },
                { name: "updatedMessage", type: "bytes" },
            ],
        },
        primaryType: "UpdateDepositDetails",
        message: {
            depositId: BigInt(depositId),
            originChainId: BigInt(originChainId),
            updatedOutputAmount,
            updatedRecipient: addressToBytes32(updatedRecipient),
            updatedMessage,
        },
    };
}
