// @coinbase/agentkit/dist/action-providers/opensea/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetNftsByAccountSchema = exports.ListNftSchema = void 0;
const zod_1 = require("zod");
/**
 * Input schema for listing an NFT on OpenSea.
 */
exports.ListNftSchema = zod_1.z
    .object({
    contractAddress: zod_1.z.string().nonempty().describe("The NFT contract address to list"),
    tokenId: zod_1.z.string().nonempty().describe("The tokenID of the NFT to list"),
    price: zod_1.z.number().positive().describe("The price in ETH to list the NFT for"),
    expirationDays: zod_1.z
        .number()
        .positive()
        .optional()
        .default(90)
        .describe("Number of days the listing should be active for (default: 90)"),
})
    .strip()
    .describe("Input schema for listing an NFT on OpenSea");
/**
 * Input schema for getting NFTs from a specific wallet address.
 */
exports.GetNftsByAccountSchema = zod_1.z
    .object({
    accountAddress: zod_1.z
        .string()
        .optional()
        .describe("The wallet address to fetch NFTs for (defaults to connected wallet if not provided)"),
})
    .strip()
    .describe("Input schema for fetching NFTs by account");
