// @zerodev/intent/_cjs/client/decorators/intent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.intentClientActions = intentClientActions;
const enableIntent_js_1 = require("../../actions/enableIntent.js");
const estimateUserIntentFees_js_1 = require("../../actions/estimateUserIntentFees.js");
const getCAB_js_1 = require("../../actions/getCAB.js");
const getIntent_js_1 = require("../../actions/getIntent.js");
const getUserIntentExecutionReceipt_js_1 = require("../../actions/getUserIntentExecutionReceipt.js");
const getUserIntentOpenReceipt_js_1 = require("../../actions/getUserIntentOpenReceipt.js");
const getUserIntentStatus_js_1 = require("../../actions/getUserIntentStatus.js");
const prepareUserIntent_js_1 = require("../../actions/prepareUserIntent.js");
const sendUserIntent_js_1 = require("../../actions/sendUserIntent.js");
const waitForUserIntentExecutionReceipt_js_1 = require("../../actions/waitForUserIntentExecutionReceipt.js");
const waitForUserIntentOpenReceipt_js_1 = require("../../actions/waitForUserIntentOpenReceipt.js");
function intentClientActions(version) {
    return (client) => ({
        getIntent: (parameters) => (0, getIntent_js_1.getIntent)(client, parameters, version),
        getCAB: (parameters) => (0, getCAB_js_1.getCAB)(client, parameters),
        prepareUserIntent: (parameters) => (0, prepareUserIntent_js_1.prepareUserIntent)(client, parameters, version),
        sendUserIntent: (parameters) => (0, sendUserIntent_js_1.sendUserIntent)(client, parameters, version),
        estimateUserIntentFees: (parameters) => (0, estimateUserIntentFees_js_1.estimateUserIntentFees)(client, parameters, version),
        getUserIntentStatus: (parameters) => (0, getUserIntentStatus_js_1.getUserIntentStatus)(client, parameters, version),
        getUserIntentOpenReceipt: (parameters) => (0, getUserIntentOpenReceipt_js_1.getUserIntentOpenReceipt)(client, parameters, version),
        getUserIntentExecutionReceipt: (parameters) => (0, getUserIntentExecutionReceipt_js_1.getUserIntentExecutionReceipt)(client, parameters, version),
        waitForUserIntentExecutionReceipt: (parameters) => (0, waitForUserIntentExecutionReceipt_js_1.waitForUserIntentExecutionReceipt)(client, parameters, version),
        waitForUserIntentOpenReceipt: (parameters) => (0, waitForUserIntentOpenReceipt_js_1.waitForUserIntentOpenReceipt)(client, parameters, version),
        enableIntent: () => (0, enableIntent_js_1.enableIntent)(client, version),
    });
}
