// @alloralabs/allora-sdk/dist/src/v2/types/cosmos/counter/v1/query.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: cosmos/counter/v1/query.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryClientImpl = exports.QueryServiceName = exports.QueryGetCountResponse = exports.QueryGetCountRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "cosmos.counter.v1";
function createBaseQueryGetCountRequest() {
    return {};
}
exports.QueryGetCountRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryGetCountRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.QueryGetCountRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseQueryGetCountRequest();
        return message;
    },
};
function createBaseQueryGetCountResponse() {
    return { totalCount: "0" };
}
exports.QueryGetCountResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.totalCount !== "0") {
            writer.uint32(8).int64(message.totalCount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryGetCountResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.totalCount = reader.int64().toString();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { totalCount: isSet(object.totalCount) ? globalThis.String(object.totalCount) : "0" };
    },
    toJSON(message) {
        const obj = {};
        if (message.totalCount !== "0") {
            obj.totalCount = message.totalCount;
        }
        return obj;
    },
    create(base) {
        return exports.QueryGetCountResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryGetCountResponse();
        message.totalCount = (_a = object.totalCount) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
exports.QueryServiceName = "cosmos.counter.v1.Query";
class QueryClientImpl {
    constructor(rpc, opts) {
        this.service = (opts === null || opts === void 0 ? void 0 : opts.service) || exports.QueryServiceName;
        this.rpc = rpc;
        this.GetCount = this.GetCount.bind(this);
    }
    GetCount(request) {
        const data = exports.QueryGetCountRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "GetCount", data);
        return promise.then((data) => exports.QueryGetCountResponse.decode(new wire_1.BinaryReader(data)));
    }
}
exports.QueryClientImpl = QueryClientImpl;
function isSet(value) {
    return value !== null && value !== undefined;
}
