// @zerodev/multi-chain-ecdsa-validator/_cjs/toMultiChainECDSAValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toMultiChainECDSAValidator = void 0;
const sdk_1 = require("@zerodev/sdk");
const account_abstraction_1 = require("viem/account-abstraction");
const accounts_1 = require("viem/accounts");
const actions_1 = require("viem/actions");
const actions_2 = require("viem/actions");
const constants_js_1 = require("./constants.js");
const ecdsaGetMultiUserOpDummySignature_js_1 = require("./utils/ecdsaGetMultiUserOpDummySignature.js");
async function toMultiChainECDSAValidator(client, { signer, entryPoint, kernelVersion: _, validatorAddress: validatorAddress_, multiChainIds }) {
    const validatorAddress = validatorAddress_ ?? constants_js_1.MULTI_CHAIN_ECDSA_VALIDATOR_ADDRESS;
    const viemSigner = await (0, sdk_1.toSigner)({ signer });
    const chainId = await (0, actions_2.getChainId)(client);
    const account = (0, accounts_1.toAccount)({
        address: viemSigner.address,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new sdk_1.SignTransactionNotSupportedBySmartAccountError();
        },
        async signTypedData(typedData) {
            return viemSigner.signTypedData(typedData);
        }
    });
    return {
        ...account,
        supportedKernelVersions: ">=0.3.0",
        validatorType: "SECONDARY",
        address: validatorAddress,
        source: "MultiChainECDSAValidator",
        getIdentifier() {
            return validatorAddress ?? constants_js_1.MULTI_CHAIN_ECDSA_VALIDATOR_ADDRESS;
        },
        async getEnableData() {
            return viemSigner.address;
        },
        async getNonceKey(_accountAddress, customNonceKey) {
            if (customNonceKey) {
                return customNonceKey;
            }
            return 0n;
        },
        async signUserOperation(userOperation) {
            const hash = (0, account_abstraction_1.getUserOperationHash)({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPointAddress: entryPoint.address,
                entryPointVersion: entryPoint.version,
                chainId: chainId
            });
            const signature = await (0, actions_1.signMessage)(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return signature;
        },
        async getStubSignature(userOperation) {
            if (!multiChainIds)
                return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
            return (0, ecdsaGetMultiUserOpDummySignature_js_1.ecdsaGetMultiUserOpDummySignature)({
                ...userOperation,
                callGasLimit: 0n,
                preVerificationGas: 0n,
                verificationGasLimit: 0n
            }, multiChainIds.length, entryPoint, chainId);
        },
        async isEnabled(_kernelAccountAddress, _selector) {
            return false;
        }
    };
}
exports.toMultiChainECDSAValidator = toMultiChainECDSAValidator;
