// @alloralabs/allora-sdk/dist/src/v2/types/generated/cosmos/protocolpool/v1/genesis.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: cosmos/protocolpool/v1/genesis.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const types_1 = require("./types");
exports.protobufPackage = "cosmos.protocolpool.v1";
function createBaseGenesisState() {
    return { continuousFund: [], budget: [], toDistribute: "" };
}
exports.GenesisState = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.continuousFund) {
            types_1.ContinuousFund.encode(v, writer.uint32(10).fork()).join();
        }
        for (const v of message.budget) {
            types_1.Budget.encode(v, writer.uint32(18).fork()).join();
        }
        if (message.toDistribute !== "") {
            writer.uint32(26).string(message.toDistribute);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.continuousFund.push(types_1.ContinuousFund.decode(reader, reader.uint32()));
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.budget.push(types_1.Budget.decode(reader, reader.uint32()));
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.toDistribute = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            continuousFund: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.continuousFund)
                ? object.continuousFund.map((e) => types_1.ContinuousFund.fromJSON(e))
                : [],
            budget: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.budget) ? object.budget.map((e) => types_1.Budget.fromJSON(e)) : [],
            toDistribute: isSet(object.toDistribute) ? globalThis.String(object.toDistribute) : "",
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if ((_a = message.continuousFund) === null || _a === void 0 ? void 0 : _a.length) {
            obj.continuousFund = message.continuousFund.map((e) => types_1.ContinuousFund.toJSON(e));
        }
        if ((_b = message.budget) === null || _b === void 0 ? void 0 : _b.length) {
            obj.budget = message.budget.map((e) => types_1.Budget.toJSON(e));
        }
        if (message.toDistribute !== "") {
            obj.toDistribute = message.toDistribute;
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseGenesisState();
        message.continuousFund = ((_a = object.continuousFund) === null || _a === void 0 ? void 0 : _a.map((e) => types_1.ContinuousFund.fromPartial(e))) || [];
        message.budget = ((_b = object.budget) === null || _b === void 0 ? void 0 : _b.map((e) => types_1.Budget.fromPartial(e))) || [];
        message.toDistribute = (_c = object.toDistribute) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
