// @coinbase/agentkit/dist/action-providers/defillama/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchProtocolsSchema = exports.GetTokenPricesSchema = exports.GetProtocolSchema = void 0;
const zod_1 = require("zod");
/**
 * Input schema for getting protocol information
 */
exports.GetProtocolSchema = zod_1.z
    .object({
    protocolId: zod_1.z.string().describe("The protocol identifier from DefiLlama"),
})
    .strict();
/**
 * Input schema for getting token prices
 */
exports.GetTokenPricesSchema = zod_1.z
    .object({
    tokens: zod_1.z
        .array(zod_1.z.string())
        .describe("Array of token addresses with chain prefix, e.g., ['ethereum:0x...']"),
    searchWidth: zod_1.z
        .string()
        .optional()
        .describe("Optional time range in minutes to search for prices, default api value is '4h'"),
})
    .strict();
/**
 * Input schema for searching protocols
 */
exports.SearchProtocolsSchema = zod_1.z
    .object({
    query: zod_1.z.string().describe("Search query to find protocols"),
})
    .strict();
