// @zerodev/intent/_esm/utils/installIntentExecutor.js
import { KernelV3_1AccountAbi } from "@zerodev/sdk";
import { concatHex, encodeAbiParameters, encodeFunctionData, parseAbiParameters, zeroAddress, } from "viem";
import { IntentVersionToAddressesMap } from "../config/constants.js";
export const installIntentExecutor = (version) => encodeFunctionData({
    abi: KernelV3_1AccountAbi,
    functionName: "installModule",
    args: [
        BigInt(2),
        IntentVersionToAddressesMap[version].intentExecutorAddress,
        concatHex([
            zeroAddress,
            encodeAbiParameters(parseAbiParameters(["bytes", "bytes"]), [
                "0x",
                "0x",
            ]),
        ]),
    ],
});
export const getInstallIntentExecutorCall = ({ accountAddress, version, }) => {
    return {
        to: accountAddress,
        data: installIntentExecutor(version),
        value: 0n,
    };
};
export const getIntentExecutorPluginData = (version) => {
    return {
        type: 2,
        address: IntentVersionToAddressesMap[version].intentExecutorAddress,
        data: concatHex([
            zeroAddress,
            encodeAbiParameters(parseAbiParameters(["bytes", "bytes"]), ["0x", "0x"]),
        ]),
    };
};
