// @alloralabs/allora-sdk/dist/src/v2/types/emissions/v5/tx.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: emissions/v5/tx.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgServiceClientImpl = exports.MsgServiceServiceName = exports.RemoveFromWhitelistAdminResponse = exports.RemoveFromWhitelistAdminRequest = exports.AddToWhitelistAdminResponse = exports.AddToWhitelistAdminRequest = exports.FundTopicResponse = exports.FundTopicRequest = exports.RewardDelegateStakeResponse = exports.RewardDelegateStakeRequest = exports.CancelRemoveDelegateStakeResponse = exports.CancelRemoveDelegateStakeRequest = exports.RemoveDelegateStakeResponse = exports.RemoveDelegateStakeRequest = exports.DelegateStakeResponse = exports.DelegateStakeRequest = exports.CancelRemoveStakeResponse = exports.CancelRemoveStakeRequest = exports.RemoveStakeResponse = exports.RemoveStakeRequest = exports.AddStakeResponse = exports.AddStakeRequest = exports.RemoveRegistrationResponse = exports.RemoveRegistrationRequest = exports.RegisterResponse = exports.RegisterRequest = exports.InsertWorkerPayloadResponse = exports.InsertWorkerPayloadRequest = exports.InsertReputerPayloadResponse = exports.InsertReputerPayloadRequest = exports.CreateNewTopicResponse = exports.CreateNewTopicRequest = exports.UpdateParamsResponse = exports.UpdateParamsRequest = exports.OptionalParams = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const reputer_1 = require("../v3/reputer");
const worker_1 = require("../v3/worker");
exports.protobufPackage = "emissions.v5";
function createBaseOptionalParams() {
    return {
        version: [],
        maxSerializedMsgLength: [],
        minTopicWeight: [],
        requiredMinimumStake: [],
        removeStakeDelayWindow: [],
        minEpochLength: [],
        betaEntropy: [],
        learningRate: [],
        maxGradientThreshold: [],
        minStakeFraction: [],
        maxUnfulfilledWorkerRequests: [],
        maxUnfulfilledReputerRequests: [],
        topicRewardStakeImportance: [],
        topicRewardFeeRevenueImportance: [],
        topicRewardAlpha: [],
        taskRewardAlpha: [],
        validatorsVsAlloraPercentReward: [],
        maxSamplesToScaleScores: [],
        maxTopInferersToReward: [],
        maxTopForecastersToReward: [],
        maxTopReputersToReward: [],
        createTopicFee: [],
        gradientDescentMaxIters: [],
        registrationFee: [],
        defaultPageLimit: [],
        maxPageLimit: [],
        minEpochLengthRecordLimit: [],
        blocksPerMonth: [],
        pRewardInference: [],
        pRewardForecast: [],
        pRewardReputer: [],
        cRewardInference: [],
        cRewardForecast: [],
        cNorm: [],
        epsilonReputer: [],
        halfMaxProcessStakeRemovalsEndBlock: [],
        dataSendingFee: [],
        epsilonSafeDiv: [],
        maxElementsPerForecast: [],
        maxActiveTopicsPerBlock: [],
        maxStringLength: [],
        initialRegretQuantile: [],
        pNormSafeDiv: [],
    };
}
exports.OptionalParams = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.version) {
            writer.uint32(10).string(v);
        }
        writer.uint32(18).fork();
        for (const v of message.maxSerializedMsgLength) {
            writer.int64(v);
        }
        writer.join();
        for (const v of message.minTopicWeight) {
            writer.uint32(26).string(v);
        }
        for (const v of message.requiredMinimumStake) {
            writer.uint32(42).string(v);
        }
        writer.uint32(50).fork();
        for (const v of message.removeStakeDelayWindow) {
            writer.int64(v);
        }
        writer.join();
        writer.uint32(58).fork();
        for (const v of message.minEpochLength) {
            writer.int64(v);
        }
        writer.join();
        for (const v of message.betaEntropy) {
            writer.uint32(66).string(v);
        }
        for (const v of message.learningRate) {
            writer.uint32(74).string(v);
        }
        for (const v of message.maxGradientThreshold) {
            writer.uint32(82).string(v);
        }
        for (const v of message.minStakeFraction) {
            writer.uint32(90).string(v);
        }
        writer.uint32(106).fork();
        for (const v of message.maxUnfulfilledWorkerRequests) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(114).fork();
        for (const v of message.maxUnfulfilledReputerRequests) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.topicRewardStakeImportance) {
            writer.uint32(122).string(v);
        }
        for (const v of message.topicRewardFeeRevenueImportance) {
            writer.uint32(130).string(v);
        }
        for (const v of message.topicRewardAlpha) {
            writer.uint32(138).string(v);
        }
        for (const v of message.taskRewardAlpha) {
            writer.uint32(146).string(v);
        }
        for (const v of message.validatorsVsAlloraPercentReward) {
            writer.uint32(154).string(v);
        }
        writer.uint32(162).fork();
        for (const v of message.maxSamplesToScaleScores) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(170).fork();
        for (const v of message.maxTopInferersToReward) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(178).fork();
        for (const v of message.maxTopForecastersToReward) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(186).fork();
        for (const v of message.maxTopReputersToReward) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.createTopicFee) {
            writer.uint32(194).string(v);
        }
        writer.uint32(202).fork();
        for (const v of message.gradientDescentMaxIters) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.registrationFee) {
            writer.uint32(226).string(v);
        }
        writer.uint32(234).fork();
        for (const v of message.defaultPageLimit) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(242).fork();
        for (const v of message.maxPageLimit) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(250).fork();
        for (const v of message.minEpochLengthRecordLimit) {
            writer.int64(v);
        }
        writer.join();
        writer.uint32(258).fork();
        for (const v of message.blocksPerMonth) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.pRewardInference) {
            writer.uint32(266).string(v);
        }
        for (const v of message.pRewardForecast) {
            writer.uint32(274).string(v);
        }
        for (const v of message.pRewardReputer) {
            writer.uint32(282).string(v);
        }
        for (const v of message.cRewardInference) {
            writer.uint32(290).string(v);
        }
        for (const v of message.cRewardForecast) {
            writer.uint32(298).string(v);
        }
        for (const v of message.cNorm) {
            writer.uint32(306).string(v);
        }
        for (const v of message.epsilonReputer) {
            writer.uint32(322).string(v);
        }
        writer.uint32(338).fork();
        for (const v of message.halfMaxProcessStakeRemovalsEndBlock) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.dataSendingFee) {
            writer.uint32(346).string(v);
        }
        for (const v of message.epsilonSafeDiv) {
            writer.uint32(354).string(v);
        }
        writer.uint32(362).fork();
        for (const v of message.maxElementsPerForecast) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(370).fork();
        for (const v of message.maxActiveTopicsPerBlock) {
            writer.uint64(v);
        }
        writer.join();
        writer.uint32(378).fork();
        for (const v of message.maxStringLength) {
            writer.uint64(v);
        }
        writer.join();
        for (const v of message.initialRegretQuantile) {
            writer.uint32(386).string(v);
        }
        for (const v of message.pNormSafeDiv) {
            writer.uint32(394).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOptionalParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.version.push(reader.string());
                    continue;
                }
                case 2: {
                    if (tag === 16) {
                        message.maxSerializedMsgLength.push(reader.int64().toString());
                        continue;
                    }
                    if (tag === 18) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxSerializedMsgLength.push(reader.int64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.minTopicWeight.push(reader.string());
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.requiredMinimumStake.push(reader.string());
                    continue;
                }
                case 6: {
                    if (tag === 48) {
                        message.removeStakeDelayWindow.push(reader.int64().toString());
                        continue;
                    }
                    if (tag === 50) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.removeStakeDelayWindow.push(reader.int64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 7: {
                    if (tag === 56) {
                        message.minEpochLength.push(reader.int64().toString());
                        continue;
                    }
                    if (tag === 58) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.minEpochLength.push(reader.int64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.betaEntropy.push(reader.string());
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.learningRate.push(reader.string());
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.maxGradientThreshold.push(reader.string());
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.minStakeFraction.push(reader.string());
                    continue;
                }
                case 13: {
                    if (tag === 104) {
                        message.maxUnfulfilledWorkerRequests.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 106) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxUnfulfilledWorkerRequests.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 14: {
                    if (tag === 112) {
                        message.maxUnfulfilledReputerRequests.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 114) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxUnfulfilledReputerRequests.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 15: {
                    if (tag !== 122) {
                        break;
                    }
                    message.topicRewardStakeImportance.push(reader.string());
                    continue;
                }
                case 16: {
                    if (tag !== 130) {
                        break;
                    }
                    message.topicRewardFeeRevenueImportance.push(reader.string());
                    continue;
                }
                case 17: {
                    if (tag !== 138) {
                        break;
                    }
                    message.topicRewardAlpha.push(reader.string());
                    continue;
                }
                case 18: {
                    if (tag !== 146) {
                        break;
                    }
                    message.taskRewardAlpha.push(reader.string());
                    continue;
                }
                case 19: {
                    if (tag !== 154) {
                        break;
                    }
                    message.validatorsVsAlloraPercentReward.push(reader.string());
                    continue;
                }
                case 20: {
                    if (tag === 160) {
                        message.maxSamplesToScaleScores.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 162) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxSamplesToScaleScores.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 21: {
                    if (tag === 168) {
                        message.maxTopInferersToReward.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 170) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxTopInferersToReward.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 22: {
                    if (tag === 176) {
                        message.maxTopForecastersToReward.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 178) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxTopForecastersToReward.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 23: {
                    if (tag === 184) {
                        message.maxTopReputersToReward.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 186) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxTopReputersToReward.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 24: {
                    if (tag !== 194) {
                        break;
                    }
                    message.createTopicFee.push(reader.string());
                    continue;
                }
                case 25: {
                    if (tag === 200) {
                        message.gradientDescentMaxIters.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 202) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.gradientDescentMaxIters.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 28: {
                    if (tag !== 226) {
                        break;
                    }
                    message.registrationFee.push(reader.string());
                    continue;
                }
                case 29: {
                    if (tag === 232) {
                        message.defaultPageLimit.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 234) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.defaultPageLimit.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 30: {
                    if (tag === 240) {
                        message.maxPageLimit.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 242) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxPageLimit.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 31: {
                    if (tag === 248) {
                        message.minEpochLengthRecordLimit.push(reader.int64().toString());
                        continue;
                    }
                    if (tag === 250) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.minEpochLengthRecordLimit.push(reader.int64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 32: {
                    if (tag === 256) {
                        message.blocksPerMonth.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 258) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.blocksPerMonth.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 33: {
                    if (tag !== 266) {
                        break;
                    }
                    message.pRewardInference.push(reader.string());
                    continue;
                }
                case 34: {
                    if (tag !== 274) {
                        break;
                    }
                    message.pRewardForecast.push(reader.string());
                    continue;
                }
                case 35: {
                    if (tag !== 282) {
                        break;
                    }
                    message.pRewardReputer.push(reader.string());
                    continue;
                }
                case 36: {
                    if (tag !== 290) {
                        break;
                    }
                    message.cRewardInference.push(reader.string());
                    continue;
                }
                case 37: {
                    if (tag !== 298) {
                        break;
                    }
                    message.cRewardForecast.push(reader.string());
                    continue;
                }
                case 38: {
                    if (tag !== 306) {
                        break;
                    }
                    message.cNorm.push(reader.string());
                    continue;
                }
                case 40: {
                    if (tag !== 322) {
                        break;
                    }
                    message.epsilonReputer.push(reader.string());
                    continue;
                }
                case 42: {
                    if (tag === 336) {
                        message.halfMaxProcessStakeRemovalsEndBlock.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 338) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.halfMaxProcessStakeRemovalsEndBlock.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 43: {
                    if (tag !== 346) {
                        break;
                    }
                    message.dataSendingFee.push(reader.string());
                    continue;
                }
                case 44: {
                    if (tag !== 354) {
                        break;
                    }
                    message.epsilonSafeDiv.push(reader.string());
                    continue;
                }
                case 45: {
                    if (tag === 360) {
                        message.maxElementsPerForecast.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 362) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxElementsPerForecast.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 46: {
                    if (tag === 368) {
                        message.maxActiveTopicsPerBlock.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 370) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxActiveTopicsPerBlock.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 47: {
                    if (tag === 376) {
                        message.maxStringLength.push(reader.uint64().toString());
                        continue;
                    }
                    if (tag === 378) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.maxStringLength.push(reader.uint64().toString());
                        }
                        continue;
                    }
                    break;
                }
                case 48: {
                    if (tag !== 386) {
                        break;
                    }
                    message.initialRegretQuantile.push(reader.string());
                    continue;
                }
                case 49: {
                    if (tag !== 394) {
                        break;
                    }
                    message.pNormSafeDiv.push(reader.string());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            version: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.version) ? object.version.map((e) => globalThis.String(e)) : [],
            maxSerializedMsgLength: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxSerializedMsgLength)
                ? object.maxSerializedMsgLength.map((e) => globalThis.String(e))
                : [],
            minTopicWeight: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.minTopicWeight)
                ? object.minTopicWeight.map((e) => globalThis.String(e))
                : [],
            requiredMinimumStake: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.requiredMinimumStake)
                ? object.requiredMinimumStake.map((e) => globalThis.String(e))
                : [],
            removeStakeDelayWindow: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.removeStakeDelayWindow)
                ? object.removeStakeDelayWindow.map((e) => globalThis.String(e))
                : [],
            minEpochLength: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.minEpochLength)
                ? object.minEpochLength.map((e) => globalThis.String(e))
                : [],
            betaEntropy: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.betaEntropy)
                ? object.betaEntropy.map((e) => globalThis.String(e))
                : [],
            learningRate: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.learningRate)
                ? object.learningRate.map((e) => globalThis.String(e))
                : [],
            maxGradientThreshold: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxGradientThreshold)
                ? object.maxGradientThreshold.map((e) => globalThis.String(e))
                : [],
            minStakeFraction: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.minStakeFraction)
                ? object.minStakeFraction.map((e) => globalThis.String(e))
                : [],
            maxUnfulfilledWorkerRequests: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxUnfulfilledWorkerRequests)
                ? object.maxUnfulfilledWorkerRequests.map((e) => globalThis.String(e))
                : [],
            maxUnfulfilledReputerRequests: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxUnfulfilledReputerRequests)
                ? object.maxUnfulfilledReputerRequests.map((e) => globalThis.String(e))
                : [],
            topicRewardStakeImportance: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.topicRewardStakeImportance)
                ? object.topicRewardStakeImportance.map((e) => globalThis.String(e))
                : [],
            topicRewardFeeRevenueImportance: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.topicRewardFeeRevenueImportance)
                ? object.topicRewardFeeRevenueImportance.map((e) => globalThis.String(e))
                : [],
            topicRewardAlpha: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.topicRewardAlpha)
                ? object.topicRewardAlpha.map((e) => globalThis.String(e))
                : [],
            taskRewardAlpha: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.taskRewardAlpha)
                ? object.taskRewardAlpha.map((e) => globalThis.String(e))
                : [],
            validatorsVsAlloraPercentReward: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.validatorsVsAlloraPercentReward)
                ? object.validatorsVsAlloraPercentReward.map((e) => globalThis.String(e))
                : [],
            maxSamplesToScaleScores: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxSamplesToScaleScores)
                ? object.maxSamplesToScaleScores.map((e) => globalThis.String(e))
                : [],
            maxTopInferersToReward: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxTopInferersToReward)
                ? object.maxTopInferersToReward.map((e) => globalThis.String(e))
                : [],
            maxTopForecastersToReward: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxTopForecastersToReward)
                ? object.maxTopForecastersToReward.map((e) => globalThis.String(e))
                : [],
            maxTopReputersToReward: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxTopReputersToReward)
                ? object.maxTopReputersToReward.map((e) => globalThis.String(e))
                : [],
            createTopicFee: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.createTopicFee)
                ? object.createTopicFee.map((e) => globalThis.String(e))
                : [],
            gradientDescentMaxIters: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.gradientDescentMaxIters)
                ? object.gradientDescentMaxIters.map((e) => globalThis.String(e))
                : [],
            registrationFee: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.registrationFee)
                ? object.registrationFee.map((e) => globalThis.String(e))
                : [],
            defaultPageLimit: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.defaultPageLimit)
                ? object.defaultPageLimit.map((e) => globalThis.String(e))
                : [],
            maxPageLimit: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxPageLimit)
                ? object.maxPageLimit.map((e) => globalThis.String(e))
                : [],
            minEpochLengthRecordLimit: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.minEpochLengthRecordLimit)
                ? object.minEpochLengthRecordLimit.map((e) => globalThis.String(e))
                : [],
            blocksPerMonth: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.blocksPerMonth)
                ? object.blocksPerMonth.map((e) => globalThis.String(e))
                : [],
            pRewardInference: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pRewardInference)
                ? object.pRewardInference.map((e) => globalThis.String(e))
                : [],
            pRewardForecast: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pRewardForecast)
                ? object.pRewardForecast.map((e) => globalThis.String(e))
                : [],
            pRewardReputer: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pRewardReputer)
                ? object.pRewardReputer.map((e) => globalThis.String(e))
                : [],
            cRewardInference: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.cRewardInference)
                ? object.cRewardInference.map((e) => globalThis.String(e))
                : [],
            cRewardForecast: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.cRewardForecast)
                ? object.cRewardForecast.map((e) => globalThis.String(e))
                : [],
            cNorm: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.cNorm) ? object.cNorm.map((e) => globalThis.String(e)) : [],
            epsilonReputer: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.epsilonReputer)
                ? object.epsilonReputer.map((e) => globalThis.String(e))
                : [],
            halfMaxProcessStakeRemovalsEndBlock: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.halfMaxProcessStakeRemovalsEndBlock)
                ? object.halfMaxProcessStakeRemovalsEndBlock.map((e) => globalThis.String(e))
                : [],
            dataSendingFee: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.dataSendingFee)
                ? object.dataSendingFee.map((e) => globalThis.String(e))
                : [],
            epsilonSafeDiv: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.epsilonSafeDiv)
                ? object.epsilonSafeDiv.map((e) => globalThis.String(e))
                : [],
            maxElementsPerForecast: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxElementsPerForecast)
                ? object.maxElementsPerForecast.map((e) => globalThis.String(e))
                : [],
            maxActiveTopicsPerBlock: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxActiveTopicsPerBlock)
                ? object.maxActiveTopicsPerBlock.map((e) => globalThis.String(e))
                : [],
            maxStringLength: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.maxStringLength)
                ? object.maxStringLength.map((e) => globalThis.String(e))
                : [],
            initialRegretQuantile: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.initialRegretQuantile)
                ? object.initialRegretQuantile.map((e) => globalThis.String(e))
                : [],
            pNormSafeDiv: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pNormSafeDiv)
                ? object.pNormSafeDiv.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
        const obj = {};
        if ((_a = message.version) === null || _a === void 0 ? void 0 : _a.length) {
            obj.version = message.version;
        }
        if ((_b = message.maxSerializedMsgLength) === null || _b === void 0 ? void 0 : _b.length) {
            obj.maxSerializedMsgLength = message.maxSerializedMsgLength;
        }
        if ((_c = message.minTopicWeight) === null || _c === void 0 ? void 0 : _c.length) {
            obj.minTopicWeight = message.minTopicWeight;
        }
        if ((_d = message.requiredMinimumStake) === null || _d === void 0 ? void 0 : _d.length) {
            obj.requiredMinimumStake = message.requiredMinimumStake;
        }
        if ((_e = message.removeStakeDelayWindow) === null || _e === void 0 ? void 0 : _e.length) {
            obj.removeStakeDelayWindow = message.removeStakeDelayWindow;
        }
        if ((_f = message.minEpochLength) === null || _f === void 0 ? void 0 : _f.length) {
            obj.minEpochLength = message.minEpochLength;
        }
        if ((_g = message.betaEntropy) === null || _g === void 0 ? void 0 : _g.length) {
            obj.betaEntropy = message.betaEntropy;
        }
        if ((_h = message.learningRate) === null || _h === void 0 ? void 0 : _h.length) {
            obj.learningRate = message.learningRate;
        }
        if ((_j = message.maxGradientThreshold) === null || _j === void 0 ? void 0 : _j.length) {
            obj.maxGradientThreshold = message.maxGradientThreshold;
        }
        if ((_k = message.minStakeFraction) === null || _k === void 0 ? void 0 : _k.length) {
            obj.minStakeFraction = message.minStakeFraction;
        }
        if ((_l = message.maxUnfulfilledWorkerRequests) === null || _l === void 0 ? void 0 : _l.length) {
            obj.maxUnfulfilledWorkerRequests = message.maxUnfulfilledWorkerRequests;
        }
        if ((_m = message.maxUnfulfilledReputerRequests) === null || _m === void 0 ? void 0 : _m.length) {
            obj.maxUnfulfilledReputerRequests = message.maxUnfulfilledReputerRequests;
        }
        if ((_o = message.topicRewardStakeImportance) === null || _o === void 0 ? void 0 : _o.length) {
            obj.topicRewardStakeImportance = message.topicRewardStakeImportance;
        }
        if ((_p = message.topicRewardFeeRevenueImportance) === null || _p === void 0 ? void 0 : _p.length) {
            obj.topicRewardFeeRevenueImportance = message.topicRewardFeeRevenueImportance;
        }
        if ((_q = message.topicRewardAlpha) === null || _q === void 0 ? void 0 : _q.length) {
            obj.topicRewardAlpha = message.topicRewardAlpha;
        }
        if ((_r = message.taskRewardAlpha) === null || _r === void 0 ? void 0 : _r.length) {
            obj.taskRewardAlpha = message.taskRewardAlpha;
        }
        if ((_s = message.validatorsVsAlloraPercentReward) === null || _s === void 0 ? void 0 : _s.length) {
            obj.validatorsVsAlloraPercentReward = message.validatorsVsAlloraPercentReward;
        }
        if ((_t = message.maxSamplesToScaleScores) === null || _t === void 0 ? void 0 : _t.length) {
            obj.maxSamplesToScaleScores = message.maxSamplesToScaleScores;
        }
        if ((_u = message.maxTopInferersToReward) === null || _u === void 0 ? void 0 : _u.length) {
            obj.maxTopInferersToReward = message.maxTopInferersToReward;
        }
        if ((_v = message.maxTopForecastersToReward) === null || _v === void 0 ? void 0 : _v.length) {
            obj.maxTopForecastersToReward = message.maxTopForecastersToReward;
        }
        if ((_w = message.maxTopReputersToReward) === null || _w === void 0 ? void 0 : _w.length) {
            obj.maxTopReputersToReward = message.maxTopReputersToReward;
        }
        if ((_x = message.createTopicFee) === null || _x === void 0 ? void 0 : _x.length) {
            obj.createTopicFee = message.createTopicFee;
        }
        if ((_y = message.gradientDescentMaxIters) === null || _y === void 0 ? void 0 : _y.length) {
            obj.gradientDescentMaxIters = message.gradientDescentMaxIters;
        }
        if ((_z = message.registrationFee) === null || _z === void 0 ? void 0 : _z.length) {
            obj.registrationFee = message.registrationFee;
        }
        if ((_0 = message.defaultPageLimit) === null || _0 === void 0 ? void 0 : _0.length) {
            obj.defaultPageLimit = message.defaultPageLimit;
        }
        if ((_1 = message.maxPageLimit) === null || _1 === void 0 ? void 0 : _1.length) {
            obj.maxPageLimit = message.maxPageLimit;
        }
        if ((_2 = message.minEpochLengthRecordLimit) === null || _2 === void 0 ? void 0 : _2.length) {
            obj.minEpochLengthRecordLimit = message.minEpochLengthRecordLimit;
        }
        if ((_3 = message.blocksPerMonth) === null || _3 === void 0 ? void 0 : _3.length) {
            obj.blocksPerMonth = message.blocksPerMonth;
        }
        if ((_4 = message.pRewardInference) === null || _4 === void 0 ? void 0 : _4.length) {
            obj.pRewardInference = message.pRewardInference;
        }
        if ((_5 = message.pRewardForecast) === null || _5 === void 0 ? void 0 : _5.length) {
            obj.pRewardForecast = message.pRewardForecast;
        }
        if ((_6 = message.pRewardReputer) === null || _6 === void 0 ? void 0 : _6.length) {
            obj.pRewardReputer = message.pRewardReputer;
        }
        if ((_7 = message.cRewardInference) === null || _7 === void 0 ? void 0 : _7.length) {
            obj.cRewardInference = message.cRewardInference;
        }
        if ((_8 = message.cRewardForecast) === null || _8 === void 0 ? void 0 : _8.length) {
            obj.cRewardForecast = message.cRewardForecast;
        }
        if ((_9 = message.cNorm) === null || _9 === void 0 ? void 0 : _9.length) {
            obj.cNorm = message.cNorm;
        }
        if ((_10 = message.epsilonReputer) === null || _10 === void 0 ? void 0 : _10.length) {
            obj.epsilonReputer = message.epsilonReputer;
        }
        if ((_11 = message.halfMaxProcessStakeRemovalsEndBlock) === null || _11 === void 0 ? void 0 : _11.length) {
            obj.halfMaxProcessStakeRemovalsEndBlock = message.halfMaxProcessStakeRemovalsEndBlock;
        }
        if ((_12 = message.dataSendingFee) === null || _12 === void 0 ? void 0 : _12.length) {
            obj.dataSendingFee = message.dataSendingFee;
        }
        if ((_13 = message.epsilonSafeDiv) === null || _13 === void 0 ? void 0 : _13.length) {
            obj.epsilonSafeDiv = message.epsilonSafeDiv;
        }
        if ((_14 = message.maxElementsPerForecast) === null || _14 === void 0 ? void 0 : _14.length) {
            obj.maxElementsPerForecast = message.maxElementsPerForecast;
        }
        if ((_15 = message.maxActiveTopicsPerBlock) === null || _15 === void 0 ? void 0 : _15.length) {
            obj.maxActiveTopicsPerBlock = message.maxActiveTopicsPerBlock;
        }
        if ((_16 = message.maxStringLength) === null || _16 === void 0 ? void 0 : _16.length) {
            obj.maxStringLength = message.maxStringLength;
        }
        if ((_17 = message.initialRegretQuantile) === null || _17 === void 0 ? void 0 : _17.length) {
            obj.initialRegretQuantile = message.initialRegretQuantile;
        }
        if ((_18 = message.pNormSafeDiv) === null || _18 === void 0 ? void 0 : _18.length) {
            obj.pNormSafeDiv = message.pNormSafeDiv;
        }
        return obj;
    },
    create(base) {
        return exports.OptionalParams.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
        const message = createBaseOptionalParams();
        message.version = ((_a = object.version) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        message.maxSerializedMsgLength = ((_b = object.maxSerializedMsgLength) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        message.minTopicWeight = ((_c = object.minTopicWeight) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        message.requiredMinimumStake = ((_d = object.requiredMinimumStake) === null || _d === void 0 ? void 0 : _d.map((e) => e)) || [];
        message.removeStakeDelayWindow = ((_e = object.removeStakeDelayWindow) === null || _e === void 0 ? void 0 : _e.map((e) => e)) || [];
        message.minEpochLength = ((_f = object.minEpochLength) === null || _f === void 0 ? void 0 : _f.map((e) => e)) || [];
        message.betaEntropy = ((_g = object.betaEntropy) === null || _g === void 0 ? void 0 : _g.map((e) => e)) || [];
        message.learningRate = ((_h = object.learningRate) === null || _h === void 0 ? void 0 : _h.map((e) => e)) || [];
        message.maxGradientThreshold = ((_j = object.maxGradientThreshold) === null || _j === void 0 ? void 0 : _j.map((e) => e)) || [];
        message.minStakeFraction = ((_k = object.minStakeFraction) === null || _k === void 0 ? void 0 : _k.map((e) => e)) || [];
        message.maxUnfulfilledWorkerRequests = ((_l = object.maxUnfulfilledWorkerRequests) === null || _l === void 0 ? void 0 : _l.map((e) => e)) || [];
        message.maxUnfulfilledReputerRequests = ((_m = object.maxUnfulfilledReputerRequests) === null || _m === void 0 ? void 0 : _m.map((e) => e)) || [];
        message.topicRewardStakeImportance = ((_o = object.topicRewardStakeImportance) === null || _o === void 0 ? void 0 : _o.map((e) => e)) || [];
        message.topicRewardFeeRevenueImportance = ((_p = object.topicRewardFeeRevenueImportance) === null || _p === void 0 ? void 0 : _p.map((e) => e)) || [];
        message.topicRewardAlpha = ((_q = object.topicRewardAlpha) === null || _q === void 0 ? void 0 : _q.map((e) => e)) || [];
        message.taskRewardAlpha = ((_r = object.taskRewardAlpha) === null || _r === void 0 ? void 0 : _r.map((e) => e)) || [];
        message.validatorsVsAlloraPercentReward = ((_s = object.validatorsVsAlloraPercentReward) === null || _s === void 0 ? void 0 : _s.map((e) => e)) || [];
        message.maxSamplesToScaleScores = ((_t = object.maxSamplesToScaleScores) === null || _t === void 0 ? void 0 : _t.map((e) => e)) || [];
        message.maxTopInferersToReward = ((_u = object.maxTopInferersToReward) === null || _u === void 0 ? void 0 : _u.map((e) => e)) || [];
        message.maxTopForecastersToReward = ((_v = object.maxTopForecastersToReward) === null || _v === void 0 ? void 0 : _v.map((e) => e)) || [];
        message.maxTopReputersToReward = ((_w = object.maxTopReputersToReward) === null || _w === void 0 ? void 0 : _w.map((e) => e)) || [];
        message.createTopicFee = ((_x = object.createTopicFee) === null || _x === void 0 ? void 0 : _x.map((e) => e)) || [];
        message.gradientDescentMaxIters = ((_y = object.gradientDescentMaxIters) === null || _y === void 0 ? void 0 : _y.map((e) => e)) || [];
        message.registrationFee = ((_z = object.registrationFee) === null || _z === void 0 ? void 0 : _z.map((e) => e)) || [];
        message.defaultPageLimit = ((_0 = object.defaultPageLimit) === null || _0 === void 0 ? void 0 : _0.map((e) => e)) || [];
        message.maxPageLimit = ((_1 = object.maxPageLimit) === null || _1 === void 0 ? void 0 : _1.map((e) => e)) || [];
        message.minEpochLengthRecordLimit = ((_2 = object.minEpochLengthRecordLimit) === null || _2 === void 0 ? void 0 : _2.map((e) => e)) || [];
        message.blocksPerMonth = ((_3 = object.blocksPerMonth) === null || _3 === void 0 ? void 0 : _3.map((e) => e)) || [];
        message.pRewardInference = ((_4 = object.pRewardInference) === null || _4 === void 0 ? void 0 : _4.map((e) => e)) || [];
        message.pRewardForecast = ((_5 = object.pRewardForecast) === null || _5 === void 0 ? void 0 : _5.map((e) => e)) || [];
        message.pRewardReputer = ((_6 = object.pRewardReputer) === null || _6 === void 0 ? void 0 : _6.map((e) => e)) || [];
        message.cRewardInference = ((_7 = object.cRewardInference) === null || _7 === void 0 ? void 0 : _7.map((e) => e)) || [];
        message.cRewardForecast = ((_8 = object.cRewardForecast) === null || _8 === void 0 ? void 0 : _8.map((e) => e)) || [];
        message.cNorm = ((_9 = object.cNorm) === null || _9 === void 0 ? void 0 : _9.map((e) => e)) || [];
        message.epsilonReputer = ((_10 = object.epsilonReputer) === null || _10 === void 0 ? void 0 : _10.map((e) => e)) || [];
        message.halfMaxProcessStakeRemovalsEndBlock = ((_11 = object.halfMaxProcessStakeRemovalsEndBlock) === null || _11 === void 0 ? void 0 : _11.map((e) => e)) || [];
        message.dataSendingFee = ((_12 = object.dataSendingFee) === null || _12 === void 0 ? void 0 : _12.map((e) => e)) || [];
        message.epsilonSafeDiv = ((_13 = object.epsilonSafeDiv) === null || _13 === void 0 ? void 0 : _13.map((e) => e)) || [];
        message.maxElementsPerForecast = ((_14 = object.maxElementsPerForecast) === null || _14 === void 0 ? void 0 : _14.map((e) => e)) || [];
        message.maxActiveTopicsPerBlock = ((_15 = object.maxActiveTopicsPerBlock) === null || _15 === void 0 ? void 0 : _15.map((e) => e)) || [];
        message.maxStringLength = ((_16 = object.maxStringLength) === null || _16 === void 0 ? void 0 : _16.map((e) => e)) || [];
        message.initialRegretQuantile = ((_17 = object.initialRegretQuantile) === null || _17 === void 0 ? void 0 : _17.map((e) => e)) || [];
        message.pNormSafeDiv = ((_18 = object.pNormSafeDiv) === null || _18 === void 0 ? void 0 : _18.map((e) => e)) || [];
        return message;
    },
};
function createBaseUpdateParamsRequest() {
    return { sender: "", params: undefined };
}
exports.UpdateParamsRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.params !== undefined) {
            exports.OptionalParams.encode(message.params, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.params = exports.OptionalParams.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            params: isSet(object.params) ? exports.OptionalParams.fromJSON(object.params) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.params !== undefined) {
            obj.params = exports.OptionalParams.toJSON(message.params);
        }
        return obj;
    },
    create(base) {
        return exports.UpdateParamsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseUpdateParamsRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.params = (object.params !== undefined && object.params !== null)
            ? exports.OptionalParams.fromPartial(object.params)
            : undefined;
        return message;
    },
};
function createBaseUpdateParamsResponse() {
    return {};
}
exports.UpdateParamsResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.UpdateParamsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseUpdateParamsResponse();
        return message;
    },
};
function createBaseCreateNewTopicRequest() {
    return {
        creator: "",
        metadata: "",
        lossMethod: "",
        epochLength: "0",
        groundTruthLag: "0",
        pNorm: "",
        alphaRegret: "",
        allowNegative: false,
        epsilon: "",
        workerSubmissionWindow: "0",
        meritSortitionAlpha: "",
        activeInfererQuantile: "",
        activeForecasterQuantile: "",
        activeReputerQuantile: "",
    };
}
exports.CreateNewTopicRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.creator !== "") {
            writer.uint32(10).string(message.creator);
        }
        if (message.metadata !== "") {
            writer.uint32(18).string(message.metadata);
        }
        if (message.lossMethod !== "") {
            writer.uint32(34).string(message.lossMethod);
        }
        if (message.epochLength !== "0") {
            writer.uint32(56).int64(message.epochLength);
        }
        if (message.groundTruthLag !== "0") {
            writer.uint32(64).int64(message.groundTruthLag);
        }
        if (message.pNorm !== "") {
            writer.uint32(82).string(message.pNorm);
        }
        if (message.alphaRegret !== "") {
            writer.uint32(90).string(message.alphaRegret);
        }
        if (message.allowNegative !== false) {
            writer.uint32(96).bool(message.allowNegative);
        }
        if (message.epsilon !== "") {
            writer.uint32(106).string(message.epsilon);
        }
        if (message.workerSubmissionWindow !== "0") {
            writer.uint32(112).int64(message.workerSubmissionWindow);
        }
        if (message.meritSortitionAlpha !== "") {
            writer.uint32(122).string(message.meritSortitionAlpha);
        }
        if (message.activeInfererQuantile !== "") {
            writer.uint32(130).string(message.activeInfererQuantile);
        }
        if (message.activeForecasterQuantile !== "") {
            writer.uint32(138).string(message.activeForecasterQuantile);
        }
        if (message.activeReputerQuantile !== "") {
            writer.uint32(146).string(message.activeReputerQuantile);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateNewTopicRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.creator = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.metadata = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.lossMethod = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.epochLength = reader.int64().toString();
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.groundTruthLag = reader.int64().toString();
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.pNorm = reader.string();
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.alphaRegret = reader.string();
                    continue;
                }
                case 12: {
                    if (tag !== 96) {
                        break;
                    }
                    message.allowNegative = reader.bool();
                    continue;
                }
                case 13: {
                    if (tag !== 106) {
                        break;
                    }
                    message.epsilon = reader.string();
                    continue;
                }
                case 14: {
                    if (tag !== 112) {
                        break;
                    }
                    message.workerSubmissionWindow = reader.int64().toString();
                    continue;
                }
                case 15: {
                    if (tag !== 122) {
                        break;
                    }
                    message.meritSortitionAlpha = reader.string();
                    continue;
                }
                case 16: {
                    if (tag !== 130) {
                        break;
                    }
                    message.activeInfererQuantile = reader.string();
                    continue;
                }
                case 17: {
                    if (tag !== 138) {
                        break;
                    }
                    message.activeForecasterQuantile = reader.string();
                    continue;
                }
                case 18: {
                    if (tag !== 146) {
                        break;
                    }
                    message.activeReputerQuantile = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            creator: isSet(object.creator) ? globalThis.String(object.creator) : "",
            metadata: isSet(object.metadata) ? globalThis.String(object.metadata) : "",
            lossMethod: isSet(object.lossMethod) ? globalThis.String(object.lossMethod) : "",
            epochLength: isSet(object.epochLength) ? globalThis.String(object.epochLength) : "0",
            groundTruthLag: isSet(object.groundTruthLag) ? globalThis.String(object.groundTruthLag) : "0",
            pNorm: isSet(object.pNorm) ? globalThis.String(object.pNorm) : "",
            alphaRegret: isSet(object.alphaRegret) ? globalThis.String(object.alphaRegret) : "",
            allowNegative: isSet(object.allowNegative) ? globalThis.Boolean(object.allowNegative) : false,
            epsilon: isSet(object.epsilon) ? globalThis.String(object.epsilon) : "",
            workerSubmissionWindow: isSet(object.workerSubmissionWindow)
                ? globalThis.String(object.workerSubmissionWindow)
                : "0",
            meritSortitionAlpha: isSet(object.meritSortitionAlpha) ? globalThis.String(object.meritSortitionAlpha) : "",
            activeInfererQuantile: isSet(object.activeInfererQuantile) ? globalThis.String(object.activeInfererQuantile) : "",
            activeForecasterQuantile: isSet(object.activeForecasterQuantile)
                ? globalThis.String(object.activeForecasterQuantile)
                : "",
            activeReputerQuantile: isSet(object.activeReputerQuantile) ? globalThis.String(object.activeReputerQuantile) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.creator !== "") {
            obj.creator = message.creator;
        }
        if (message.metadata !== "") {
            obj.metadata = message.metadata;
        }
        if (message.lossMethod !== "") {
            obj.lossMethod = message.lossMethod;
        }
        if (message.epochLength !== "0") {
            obj.epochLength = message.epochLength;
        }
        if (message.groundTruthLag !== "0") {
            obj.groundTruthLag = message.groundTruthLag;
        }
        if (message.pNorm !== "") {
            obj.pNorm = message.pNorm;
        }
        if (message.alphaRegret !== "") {
            obj.alphaRegret = message.alphaRegret;
        }
        if (message.allowNegative !== false) {
            obj.allowNegative = message.allowNegative;
        }
        if (message.epsilon !== "") {
            obj.epsilon = message.epsilon;
        }
        if (message.workerSubmissionWindow !== "0") {
            obj.workerSubmissionWindow = message.workerSubmissionWindow;
        }
        if (message.meritSortitionAlpha !== "") {
            obj.meritSortitionAlpha = message.meritSortitionAlpha;
        }
        if (message.activeInfererQuantile !== "") {
            obj.activeInfererQuantile = message.activeInfererQuantile;
        }
        if (message.activeForecasterQuantile !== "") {
            obj.activeForecasterQuantile = message.activeForecasterQuantile;
        }
        if (message.activeReputerQuantile !== "") {
            obj.activeReputerQuantile = message.activeReputerQuantile;
        }
        return obj;
    },
    create(base) {
        return exports.CreateNewTopicRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        const message = createBaseCreateNewTopicRequest();
        message.creator = (_a = object.creator) !== null && _a !== void 0 ? _a : "";
        message.metadata = (_b = object.metadata) !== null && _b !== void 0 ? _b : "";
        message.lossMethod = (_c = object.lossMethod) !== null && _c !== void 0 ? _c : "";
        message.epochLength = (_d = object.epochLength) !== null && _d !== void 0 ? _d : "0";
        message.groundTruthLag = (_e = object.groundTruthLag) !== null && _e !== void 0 ? _e : "0";
        message.pNorm = (_f = object.pNorm) !== null && _f !== void 0 ? _f : "";
        message.alphaRegret = (_g = object.alphaRegret) !== null && _g !== void 0 ? _g : "";
        message.allowNegative = (_h = object.allowNegative) !== null && _h !== void 0 ? _h : false;
        message.epsilon = (_j = object.epsilon) !== null && _j !== void 0 ? _j : "";
        message.workerSubmissionWindow = (_k = object.workerSubmissionWindow) !== null && _k !== void 0 ? _k : "0";
        message.meritSortitionAlpha = (_l = object.meritSortitionAlpha) !== null && _l !== void 0 ? _l : "";
        message.activeInfererQuantile = (_m = object.activeInfererQuantile) !== null && _m !== void 0 ? _m : "";
        message.activeForecasterQuantile = (_o = object.activeForecasterQuantile) !== null && _o !== void 0 ? _o : "";
        message.activeReputerQuantile = (_p = object.activeReputerQuantile) !== null && _p !== void 0 ? _p : "";
        return message;
    },
};
function createBaseCreateNewTopicResponse() {
    return { topicId: "0" };
}
exports.CreateNewTopicResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.topicId !== "0") {
            writer.uint32(8).uint64(message.topicId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCreateNewTopicResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0" };
    },
    toJSON(message) {
        const obj = {};
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        return obj;
    },
    create(base) {
        return exports.CreateNewTopicResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseCreateNewTopicResponse();
        message.topicId = (_a = object.topicId) !== null && _a !== void 0 ? _a : "0";
        return message;
    },
};
function createBaseInsertReputerPayloadRequest() {
    return { sender: "", reputerValueBundle: undefined };
}
exports.InsertReputerPayloadRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.reputerValueBundle !== undefined) {
            reputer_1.ReputerValueBundle.encode(message.reputerValueBundle, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInsertReputerPayloadRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.reputerValueBundle = reputer_1.ReputerValueBundle.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            reputerValueBundle: isSet(object.reputerValueBundle)
                ? reputer_1.ReputerValueBundle.fromJSON(object.reputerValueBundle)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.reputerValueBundle !== undefined) {
            obj.reputerValueBundle = reputer_1.ReputerValueBundle.toJSON(message.reputerValueBundle);
        }
        return obj;
    },
    create(base) {
        return exports.InsertReputerPayloadRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseInsertReputerPayloadRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.reputerValueBundle = (object.reputerValueBundle !== undefined && object.reputerValueBundle !== null)
            ? reputer_1.ReputerValueBundle.fromPartial(object.reputerValueBundle)
            : undefined;
        return message;
    },
};
function createBaseInsertReputerPayloadResponse() {
    return {};
}
exports.InsertReputerPayloadResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInsertReputerPayloadResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.InsertReputerPayloadResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseInsertReputerPayloadResponse();
        return message;
    },
};
function createBaseInsertWorkerPayloadRequest() {
    return { sender: "", workerDataBundle: undefined };
}
exports.InsertWorkerPayloadRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.workerDataBundle !== undefined) {
            worker_1.WorkerDataBundle.encode(message.workerDataBundle, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInsertWorkerPayloadRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.workerDataBundle = worker_1.WorkerDataBundle.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            workerDataBundle: isSet(object.workerDataBundle) ? worker_1.WorkerDataBundle.fromJSON(object.workerDataBundle) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.workerDataBundle !== undefined) {
            obj.workerDataBundle = worker_1.WorkerDataBundle.toJSON(message.workerDataBundle);
        }
        return obj;
    },
    create(base) {
        return exports.InsertWorkerPayloadRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseInsertWorkerPayloadRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.workerDataBundle = (object.workerDataBundle !== undefined && object.workerDataBundle !== null)
            ? worker_1.WorkerDataBundle.fromPartial(object.workerDataBundle)
            : undefined;
        return message;
    },
};
function createBaseInsertWorkerPayloadResponse() {
    return {};
}
exports.InsertWorkerPayloadResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInsertWorkerPayloadResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.InsertWorkerPayloadResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseInsertWorkerPayloadResponse();
        return message;
    },
};
function createBaseRegisterRequest() {
    return { sender: "", topicId: "0", owner: "", isReputer: false };
}
exports.RegisterRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(32).uint64(message.topicId);
        }
        if (message.owner !== "") {
            writer.uint32(42).string(message.owner);
        }
        if (message.isReputer !== false) {
            writer.uint32(48).bool(message.isReputer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRegisterRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.owner = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.isReputer = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            owner: isSet(object.owner) ? globalThis.String(object.owner) : "",
            isReputer: isSet(object.isReputer) ? globalThis.Boolean(object.isReputer) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.owner !== "") {
            obj.owner = message.owner;
        }
        if (message.isReputer !== false) {
            obj.isReputer = message.isReputer;
        }
        return obj;
    },
    create(base) {
        return exports.RegisterRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseRegisterRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.owner = (_c = object.owner) !== null && _c !== void 0 ? _c : "";
        message.isReputer = (_d = object.isReputer) !== null && _d !== void 0 ? _d : false;
        return message;
    },
};
function createBaseRegisterResponse() {
    return { success: false, message: "" };
}
exports.RegisterResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.message !== "") {
            writer.uint32(18).string(message.message);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRegisterResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            message: isSet(object.message) ? globalThis.String(object.message) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.message !== "") {
            obj.message = message.message;
        }
        return obj;
    },
    create(base) {
        return exports.RegisterResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseRegisterResponse();
        message.success = (_a = object.success) !== null && _a !== void 0 ? _a : false;
        message.message = (_b = object.message) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseRemoveRegistrationRequest() {
    return { sender: "", topicId: "0", isReputer: false };
}
exports.RemoveRegistrationRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.isReputer !== false) {
            writer.uint32(24).bool(message.isReputer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveRegistrationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.isReputer = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            isReputer: isSet(object.isReputer) ? globalThis.Boolean(object.isReputer) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.isReputer !== false) {
            obj.isReputer = message.isReputer;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveRegistrationRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseRemoveRegistrationRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.isReputer = (_c = object.isReputer) !== null && _c !== void 0 ? _c : false;
        return message;
    },
};
function createBaseRemoveRegistrationResponse() {
    return { success: false, message: "" };
}
exports.RemoveRegistrationResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.message !== "") {
            writer.uint32(18).string(message.message);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveRegistrationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            message: isSet(object.message) ? globalThis.String(object.message) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.message !== "") {
            obj.message = message.message;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveRegistrationResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseRemoveRegistrationResponse();
        message.success = (_a = object.success) !== null && _a !== void 0 ? _a : false;
        message.message = (_b = object.message) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseAddStakeRequest() {
    return { sender: "", topicId: "0", amount: "" };
}
exports.AddStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.amount !== "") {
            writer.uint32(26).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.amount = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            amount: isSet(object.amount) ? globalThis.String(object.amount) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.amount !== "") {
            obj.amount = message.amount;
        }
        return obj;
    },
    create(base) {
        return exports.AddStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseAddStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.amount = (_c = object.amount) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseAddStakeResponse() {
    return {};
}
exports.AddStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.AddStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseAddStakeResponse();
        return message;
    },
};
function createBaseRemoveStakeRequest() {
    return { sender: "", topicId: "0", amount: "" };
}
exports.RemoveStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.amount !== "") {
            writer.uint32(26).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.amount = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            amount: isSet(object.amount) ? globalThis.String(object.amount) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.amount !== "") {
            obj.amount = message.amount;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseRemoveStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.amount = (_c = object.amount) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseRemoveStakeResponse() {
    return {};
}
exports.RemoveStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RemoveStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseRemoveStakeResponse();
        return message;
    },
};
function createBaseCancelRemoveStakeRequest() {
    return { sender: "", topicId: "0" };
}
exports.CancelRemoveStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCancelRemoveStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        return obj;
    },
    create(base) {
        return exports.CancelRemoveStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseCancelRemoveStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        return message;
    },
};
function createBaseCancelRemoveStakeResponse() {
    return {};
}
exports.CancelRemoveStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCancelRemoveStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.CancelRemoveStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseCancelRemoveStakeResponse();
        return message;
    },
};
function createBaseDelegateStakeRequest() {
    return { sender: "", topicId: "0", reputer: "", amount: "" };
}
exports.DelegateStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.reputer !== "") {
            writer.uint32(26).string(message.reputer);
        }
        if (message.amount !== "") {
            writer.uint32(34).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDelegateStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.reputer = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.amount = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            reputer: isSet(object.reputer) ? globalThis.String(object.reputer) : "",
            amount: isSet(object.amount) ? globalThis.String(object.amount) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.reputer !== "") {
            obj.reputer = message.reputer;
        }
        if (message.amount !== "") {
            obj.amount = message.amount;
        }
        return obj;
    },
    create(base) {
        return exports.DelegateStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseDelegateStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.reputer = (_c = object.reputer) !== null && _c !== void 0 ? _c : "";
        message.amount = (_d = object.amount) !== null && _d !== void 0 ? _d : "";
        return message;
    },
};
function createBaseDelegateStakeResponse() {
    return {};
}
exports.DelegateStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDelegateStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.DelegateStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseDelegateStakeResponse();
        return message;
    },
};
function createBaseRemoveDelegateStakeRequest() {
    return { sender: "", reputer: "", topicId: "0", amount: "" };
}
exports.RemoveDelegateStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.reputer !== "") {
            writer.uint32(18).string(message.reputer);
        }
        if (message.topicId !== "0") {
            writer.uint32(24).uint64(message.topicId);
        }
        if (message.amount !== "") {
            writer.uint32(34).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveDelegateStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.reputer = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.amount = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            reputer: isSet(object.reputer) ? globalThis.String(object.reputer) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            amount: isSet(object.amount) ? globalThis.String(object.amount) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.reputer !== "") {
            obj.reputer = message.reputer;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.amount !== "") {
            obj.amount = message.amount;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveDelegateStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseRemoveDelegateStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.reputer = (_b = object.reputer) !== null && _b !== void 0 ? _b : "";
        message.topicId = (_c = object.topicId) !== null && _c !== void 0 ? _c : "0";
        message.amount = (_d = object.amount) !== null && _d !== void 0 ? _d : "";
        return message;
    },
};
function createBaseRemoveDelegateStakeResponse() {
    return {};
}
exports.RemoveDelegateStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveDelegateStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RemoveDelegateStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseRemoveDelegateStakeResponse();
        return message;
    },
};
function createBaseCancelRemoveDelegateStakeRequest() {
    return { sender: "", topicId: "0", delegator: "", reputer: "" };
}
exports.CancelRemoveDelegateStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.delegator !== "") {
            writer.uint32(26).string(message.delegator);
        }
        if (message.reputer !== "") {
            writer.uint32(34).string(message.reputer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCancelRemoveDelegateStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.delegator = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.reputer = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            delegator: isSet(object.delegator) ? globalThis.String(object.delegator) : "",
            reputer: isSet(object.reputer) ? globalThis.String(object.reputer) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.delegator !== "") {
            obj.delegator = message.delegator;
        }
        if (message.reputer !== "") {
            obj.reputer = message.reputer;
        }
        return obj;
    },
    create(base) {
        return exports.CancelRemoveDelegateStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseCancelRemoveDelegateStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.delegator = (_c = object.delegator) !== null && _c !== void 0 ? _c : "";
        message.reputer = (_d = object.reputer) !== null && _d !== void 0 ? _d : "";
        return message;
    },
};
function createBaseCancelRemoveDelegateStakeResponse() {
    return {};
}
exports.CancelRemoveDelegateStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCancelRemoveDelegateStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.CancelRemoveDelegateStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseCancelRemoveDelegateStakeResponse();
        return message;
    },
};
function createBaseRewardDelegateStakeRequest() {
    return { sender: "", topicId: "0", reputer: "" };
}
exports.RewardDelegateStakeRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.reputer !== "") {
            writer.uint32(26).string(message.reputer);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRewardDelegateStakeRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.reputer = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            reputer: isSet(object.reputer) ? globalThis.String(object.reputer) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.reputer !== "") {
            obj.reputer = message.reputer;
        }
        return obj;
    },
    create(base) {
        return exports.RewardDelegateStakeRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseRewardDelegateStakeRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.reputer = (_c = object.reputer) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseRewardDelegateStakeResponse() {
    return {};
}
exports.RewardDelegateStakeResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRewardDelegateStakeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RewardDelegateStakeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseRewardDelegateStakeResponse();
        return message;
    },
};
function createBaseFundTopicRequest() {
    return { sender: "", topicId: "0", amount: "" };
}
exports.FundTopicRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.topicId !== "0") {
            writer.uint32(16).uint64(message.topicId);
        }
        if (message.amount !== "") {
            writer.uint32(26).string(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFundTopicRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.topicId = reader.uint64().toString();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.amount = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            topicId: isSet(object.topicId) ? globalThis.String(object.topicId) : "0",
            amount: isSet(object.amount) ? globalThis.String(object.amount) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.topicId !== "0") {
            obj.topicId = message.topicId;
        }
        if (message.amount !== "") {
            obj.amount = message.amount;
        }
        return obj;
    },
    create(base) {
        return exports.FundTopicRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseFundTopicRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.topicId = (_b = object.topicId) !== null && _b !== void 0 ? _b : "0";
        message.amount = (_c = object.amount) !== null && _c !== void 0 ? _c : "";
        return message;
    },
};
function createBaseFundTopicResponse() {
    return {};
}
exports.FundTopicResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFundTopicResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.FundTopicResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseFundTopicResponse();
        return message;
    },
};
function createBaseAddToWhitelistAdminRequest() {
    return { sender: "", address: "" };
}
exports.AddToWhitelistAdminRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.address !== "") {
            writer.uint32(18).string(message.address);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddToWhitelistAdminRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            address: isSet(object.address) ? globalThis.String(object.address) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.address !== "") {
            obj.address = message.address;
        }
        return obj;
    },
    create(base) {
        return exports.AddToWhitelistAdminRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseAddToWhitelistAdminRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.address = (_b = object.address) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseAddToWhitelistAdminResponse() {
    return {};
}
exports.AddToWhitelistAdminResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAddToWhitelistAdminResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.AddToWhitelistAdminResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseAddToWhitelistAdminResponse();
        return message;
    },
};
function createBaseRemoveFromWhitelistAdminRequest() {
    return { sender: "", address: "" };
}
exports.RemoveFromWhitelistAdminRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.address !== "") {
            writer.uint32(18).string(message.address);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveFromWhitelistAdminRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.sender = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            sender: isSet(object.sender) ? globalThis.String(object.sender) : "",
            address: isSet(object.address) ? globalThis.String(object.address) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.sender !== "") {
            obj.sender = message.sender;
        }
        if (message.address !== "") {
            obj.address = message.address;
        }
        return obj;
    },
    create(base) {
        return exports.RemoveFromWhitelistAdminRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseRemoveFromWhitelistAdminRequest();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.address = (_b = object.address) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseRemoveFromWhitelistAdminResponse() {
    return {};
}
exports.RemoveFromWhitelistAdminResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRemoveFromWhitelistAdminResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.RemoveFromWhitelistAdminResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseRemoveFromWhitelistAdminResponse();
        return message;
    },
};
exports.MsgServiceServiceName = "emissions.v5.MsgService";
class MsgServiceClientImpl {
    constructor(rpc, opts) {
        this.service = (opts === null || opts === void 0 ? void 0 : opts.service) || exports.MsgServiceServiceName;
        this.rpc = rpc;
        this.UpdateParams = this.UpdateParams.bind(this);
        this.CreateNewTopic = this.CreateNewTopic.bind(this);
        this.Register = this.Register.bind(this);
        this.RemoveRegistration = this.RemoveRegistration.bind(this);
        this.AddStake = this.AddStake.bind(this);
        this.RemoveStake = this.RemoveStake.bind(this);
        this.CancelRemoveStake = this.CancelRemoveStake.bind(this);
        this.DelegateStake = this.DelegateStake.bind(this);
        this.RewardDelegateStake = this.RewardDelegateStake.bind(this);
        this.RemoveDelegateStake = this.RemoveDelegateStake.bind(this);
        this.CancelRemoveDelegateStake = this.CancelRemoveDelegateStake.bind(this);
        this.FundTopic = this.FundTopic.bind(this);
        this.AddToWhitelistAdmin = this.AddToWhitelistAdmin.bind(this);
        this.RemoveFromWhitelistAdmin = this.RemoveFromWhitelistAdmin.bind(this);
        this.InsertWorkerPayload = this.InsertWorkerPayload.bind(this);
        this.InsertReputerPayload = this.InsertReputerPayload.bind(this);
    }
    UpdateParams(request) {
        const data = exports.UpdateParamsRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "UpdateParams", data);
        return promise.then((data) => exports.UpdateParamsResponse.decode(new wire_1.BinaryReader(data)));
    }
    CreateNewTopic(request) {
        const data = exports.CreateNewTopicRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CreateNewTopic", data);
        return promise.then((data) => exports.CreateNewTopicResponse.decode(new wire_1.BinaryReader(data)));
    }
    Register(request) {
        const data = exports.RegisterRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "Register", data);
        return promise.then((data) => exports.RegisterResponse.decode(new wire_1.BinaryReader(data)));
    }
    RemoveRegistration(request) {
        const data = exports.RemoveRegistrationRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RemoveRegistration", data);
        return promise.then((data) => exports.RemoveRegistrationResponse.decode(new wire_1.BinaryReader(data)));
    }
    AddStake(request) {
        const data = exports.AddStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AddStake", data);
        return promise.then((data) => exports.AddStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    RemoveStake(request) {
        const data = exports.RemoveStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RemoveStake", data);
        return promise.then((data) => exports.RemoveStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    CancelRemoveStake(request) {
        const data = exports.CancelRemoveStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CancelRemoveStake", data);
        return promise.then((data) => exports.CancelRemoveStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    DelegateStake(request) {
        const data = exports.DelegateStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "DelegateStake", data);
        return promise.then((data) => exports.DelegateStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    RewardDelegateStake(request) {
        const data = exports.RewardDelegateStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RewardDelegateStake", data);
        return promise.then((data) => exports.RewardDelegateStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    RemoveDelegateStake(request) {
        const data = exports.RemoveDelegateStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RemoveDelegateStake", data);
        return promise.then((data) => exports.RemoveDelegateStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    CancelRemoveDelegateStake(request) {
        const data = exports.CancelRemoveDelegateStakeRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "CancelRemoveDelegateStake", data);
        return promise.then((data) => exports.CancelRemoveDelegateStakeResponse.decode(new wire_1.BinaryReader(data)));
    }
    FundTopic(request) {
        const data = exports.FundTopicRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "FundTopic", data);
        return promise.then((data) => exports.FundTopicResponse.decode(new wire_1.BinaryReader(data)));
    }
    AddToWhitelistAdmin(request) {
        const data = exports.AddToWhitelistAdminRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "AddToWhitelistAdmin", data);
        return promise.then((data) => exports.AddToWhitelistAdminResponse.decode(new wire_1.BinaryReader(data)));
    }
    RemoveFromWhitelistAdmin(request) {
        const data = exports.RemoveFromWhitelistAdminRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "RemoveFromWhitelistAdmin", data);
        return promise.then((data) => exports.RemoveFromWhitelistAdminResponse.decode(new wire_1.BinaryReader(data)));
    }
    InsertWorkerPayload(request) {
        const data = exports.InsertWorkerPayloadRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "InsertWorkerPayload", data);
        return promise.then((data) => exports.InsertWorkerPayloadResponse.decode(new wire_1.BinaryReader(data)));
    }
    InsertReputerPayload(request) {
        const data = exports.InsertReputerPayloadRequest.encode(request).finish();
        const promise = this.rpc.request(this.service, "InsertReputerPayload", data);
        return promise.then((data) => exports.InsertReputerPayloadResponse.decode(new wire_1.BinaryReader(data)));
    }
}
exports.MsgServiceClientImpl = MsgServiceClientImpl;
function isSet(value) {
    return value !== null && value !== undefined;
}
