// @xmtp/content-type-wallet-send-calls/dist/index.js
import { ContentTypeId } from '@xmtp/content-type-primitives';

/* eslint-disable @typescript-eslint/no-explicit-any */
const ContentTypeWalletSendCalls = new ContentTypeId({
    authorityId: "xmtp.org",
    typeId: "walletSendCalls",
    versionMajor: 1,
    versionMinor: 0,
});
class WalletSendCallsCodec {
    get contentType() {
        return ContentTypeWalletSendCalls;
    }
    encode(content) {
        const encoded = {
            type: ContentTypeWalletSendCalls,
            parameters: {},
            content: new TextEncoder().encode(JSON.stringify(content)),
        };
        return encoded;
    }
    decode(encodedContent) {
        const uint8Array = encodedContent.content;
        const contentReceived = JSON.parse(new TextDecoder().decode(uint8Array));
        return contentReceived;
    }
    fallback(content) {
        return `[Transaction request generated]: ${JSON.stringify(content)}`;
    }
    shouldPush() {
        return true;
    }
}

export { ContentTypeWalletSendCalls, WalletSendCallsCodec };
