// @xmtp/proto/ts/dist/cjs/device_sync/event_backup.pb.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSave = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "xmtp.device_sync.event_backup";
function createBaseEventSave() {
    return {
        createdAtNs: long_1.default.ZERO,
        event: "",
        details: new Uint8Array(),
        groupId: undefined,
    };
}
exports.EventSave = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (!message.createdAtNs.isZero()) {
            writer.uint32(8).int64(message.createdAtNs);
        }
        if (message.event !== "") {
            writer.uint32(18).string(message.event);
        }
        if (message.details.length !== 0) {
            writer.uint32(26).bytes(message.details);
        }
        if (message.groupId !== undefined) {
            writer.uint32(34).bytes(message.groupId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEventSave();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.createdAtNs = reader.int64();
                    break;
                case 2:
                    message.event = reader.string();
                    break;
                case 3:
                    message.details = reader.bytes();
                    break;
                case 4:
                    message.groupId = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            createdAtNs: isSet(object.createdAtNs)
                ? long_1.default.fromValue(object.createdAtNs)
                : long_1.default.ZERO,
            event: isSet(object.event) ? String(object.event) : "",
            details: isSet(object.details)
                ? bytesFromBase64(object.details)
                : new Uint8Array(),
            groupId: isSet(object.groupId)
                ? bytesFromBase64(object.groupId)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        message.createdAtNs !== undefined &&
            (obj.createdAtNs = (message.createdAtNs || long_1.default.ZERO).toString());
        message.event !== undefined && (obj.event = message.event);
        message.details !== undefined &&
            (obj.details = base64FromBytes(message.details !== undefined ? message.details : new Uint8Array()));
        message.groupId !== undefined &&
            (obj.groupId =
                message.groupId !== undefined
                    ? base64FromBytes(message.groupId)
                    : undefined);
        return obj;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseEventSave();
        message.createdAtNs =
            object.createdAtNs !== undefined && object.createdAtNs !== null
                ? long_1.default.fromValue(object.createdAtNs)
                : long_1.default.ZERO;
        message.event = (_a = object.event) !== null && _a !== void 0 ? _a : "";
        message.details = (_b = object.details) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.groupId = (_c = object.groupId) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
var globalThis = (() => {
    if (typeof globalThis !== "undefined")
        return globalThis;
    if (typeof self !== "undefined")
        return self;
    if (typeof window !== "undefined")
        return window;
    if (typeof global !== "undefined")
        return global;
    throw "Unable to locate global object";
})();
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
