// @alloralabs/allora-sdk/dist/src/v2/types/generated/emissions/v2/node.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.3.0
//   protoc               unknown
// source: emissions/v2/node.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.OffchainNode = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "emissions.v2";
function createBaseOffchainNode() {
    return { owner: "", nodeAddress: "" };
}
exports.OffchainNode = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.owner !== "") {
            writer.uint32(26).string(message.owner);
        }
        if (message.nodeAddress !== "") {
            writer.uint32(34).string(message.nodeAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseOffchainNode();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.owner = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.nodeAddress = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            owner: isSet(object.owner) ? globalThis.String(object.owner) : "",
            nodeAddress: isSet(object.nodeAddress) ? globalThis.String(object.nodeAddress) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.owner !== "") {
            obj.owner = message.owner;
        }
        if (message.nodeAddress !== "") {
            obj.nodeAddress = message.nodeAddress;
        }
        return obj;
    },
    create(base) {
        return exports.OffchainNode.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseOffchainNode();
        message.owner = (_a = object.owner) !== null && _a !== void 0 ? _a : "";
        message.nodeAddress = (_b = object.nodeAddress) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
