// @zerodev/intent/_esm/client/decorators/intent.js
import { enableIntent, } from "../../actions/enableIntent.js";
import { estimateUserIntentFees, } from "../../actions/estimateUserIntentFees.js";
import { getCAB, } from "../../actions/getCAB.js";
import { getIntent, } from "../../actions/getIntent.js";
import { getUserIntentExecutionReceipt } from "../../actions/getUserIntentExecutionReceipt.js";
import { getUserIntentOpenReceipt } from "../../actions/getUserIntentOpenReceipt.js";
import { getUserIntentStatus, } from "../../actions/getUserIntentStatus.js";
import { prepareUserIntent, } from "../../actions/prepareUserIntent.js";
import { sendUserIntent, } from "../../actions/sendUserIntent.js";
import { waitForUserIntentExecutionReceipt, } from "../../actions/waitForUserIntentExecutionReceipt.js";
import { waitForUserIntentOpenReceipt, } from "../../actions/waitForUserIntentOpenReceipt.js";
export function intentClientActions(version) {
    return (client) => ({
        getIntent: (parameters) => getIntent(client, parameters, version),
        getCAB: (parameters) => getCAB(client, parameters),
        prepareUserIntent: (parameters) => prepareUserIntent(client, parameters, version),
        sendUserIntent: (parameters) => sendUserIntent(client, parameters, version),
        estimateUserIntentFees: (parameters) => estimateUserIntentFees(client, parameters, version),
        getUserIntentStatus: (parameters) => getUserIntentStatus(client, parameters, version),
        getUserIntentOpenReceipt: (parameters) => getUserIntentOpenReceipt(client, parameters, version),
        getUserIntentExecutionReceipt: (parameters) => getUserIntentExecutionReceipt(client, parameters, version),
        waitForUserIntentExecutionReceipt: (parameters) => waitForUserIntentExecutionReceipt(client, parameters, version),
        waitForUserIntentOpenReceipt: (parameters) => waitForUserIntentOpenReceipt(client, parameters, version),
        enableIntent: () => enableIntent(client, version),
    });
}
